--- reference/api/typescript.d.ts
+++ local/api/typescript.d.ts
@@ -4490,10 +4490,10 @@
         _declarationBrand: undefined;
     }
     interface NamedDeclaration extends Declaration {
-        readonly name?: DeclarationName;
+        readonly name?: DeclarationName | undefined;
     }
     interface DeclarationStatement extends NamedDeclaration, Statement {
-        readonly name?: Identifier | StringLiteral | NumericLiteral;
+        readonly name?: Identifier | StringLiteral | NumericLiteral | undefined;
     }
     interface ComputedPropertyName extends Node {
         readonly kind: SyntaxKind.ComputedPropertyName;
@@ -4515,16 +4515,16 @@
     interface TypeParameterDeclaration extends NamedDeclaration, JSDocContainer {
         readonly kind: SyntaxKind.TypeParameter;
         readonly parent: DeclarationWithTypeParameterChildren | InferTypeNode;
-        readonly modifiers?: NodeArray<Modifier>;
+        readonly modifiers?: NodeArray<Modifier> | undefined;
         readonly name: Identifier;
         /** Note: Consider calling `getEffectiveConstraintOfTypeParameter` */
-        readonly constraint?: TypeNode;
-        readonly default?: TypeNode;
-        expression?: Expression;
+        readonly constraint?: TypeNode | undefined;
+        readonly default?: TypeNode | undefined;
+        expression?: Expression | undefined;
     }
     interface SignatureDeclarationBase extends NamedDeclaration, JSDocContainer {
         readonly kind: SignatureDeclaration["kind"];
-        readonly name?: PropertyName;
+        readonly name?: PropertyName | undefined;
         readonly typeParameters?: NodeArray<TypeParameterDeclaration> | undefined;
         readonly parameters: NodeArray<ParameterDeclaration>;
         readonly type?: TypeNode | undefined;
@@ -4541,9 +4541,9 @@
         readonly kind: SyntaxKind.VariableDeclaration;
         readonly parent: VariableDeclarationList | CatchClause;
         readonly name: BindingName;
-        readonly exclamationToken?: ExclamationToken;
-        readonly type?: TypeNode;
-        readonly initializer?: Expression;
+        readonly exclamationToken?: ExclamationToken | undefined;
+        readonly type?: TypeNode | undefined;
+        readonly initializer?: Expression | undefined;
     }
     interface VariableDeclarationList extends Node {
         readonly kind: SyntaxKind.VariableDeclarationList;
@@ -4553,45 +4553,45 @@
     interface ParameterDeclaration extends NamedDeclaration, JSDocContainer {
         readonly kind: SyntaxKind.Parameter;
         readonly parent: SignatureDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
-        readonly dotDotDotToken?: DotDotDotToken;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
+        readonly dotDotDotToken?: DotDotDotToken | undefined;
         readonly name: BindingName;
-        readonly questionToken?: QuestionToken;
-        readonly type?: TypeNode;
-        readonly initializer?: Expression;
+        readonly questionToken?: QuestionToken | undefined;
+        readonly type?: TypeNode | undefined;
+        readonly initializer?: Expression | undefined;
     }
     interface BindingElement extends NamedDeclaration, FlowContainer {
         readonly kind: SyntaxKind.BindingElement;
         readonly parent: BindingPattern;
-        readonly propertyName?: PropertyName;
-        readonly dotDotDotToken?: DotDotDotToken;
+        readonly propertyName?: PropertyName | undefined;
+        readonly dotDotDotToken?: DotDotDotToken | undefined;
         readonly name: BindingName;
-        readonly initializer?: Expression;
+        readonly initializer?: Expression | undefined;
     }
     interface PropertySignature extends TypeElement, JSDocContainer {
         readonly kind: SyntaxKind.PropertySignature;
         readonly parent: TypeLiteralNode | InterfaceDeclaration;
-        readonly modifiers?: NodeArray<Modifier>;
+        readonly modifiers?: NodeArray<Modifier> | undefined;
         readonly name: PropertyName;
-        readonly questionToken?: QuestionToken;
-        readonly type?: TypeNode;
+        readonly questionToken?: QuestionToken | undefined;
+        readonly type?: TypeNode | undefined;
     }
     interface PropertyDeclaration extends ClassElement, JSDocContainer {
         readonly kind: SyntaxKind.PropertyDeclaration;
         readonly parent: ClassLikeDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly name: PropertyName;
-        readonly questionToken?: QuestionToken;
-        readonly exclamationToken?: ExclamationToken;
-        readonly type?: TypeNode;
-        readonly initializer?: Expression;
+        readonly questionToken?: QuestionToken | undefined;
+        readonly exclamationToken?: ExclamationToken | undefined;
+        readonly type?: TypeNode | undefined;
+        readonly initializer?: Expression | undefined;
     }
     interface AutoAccessorPropertyDeclaration extends PropertyDeclaration {
         _autoAccessorBrand: undefined;
     }
     interface ObjectLiteralElement extends NamedDeclaration {
         _objectLiteralBrand: undefined;
-        readonly name?: PropertyName;
+        readonly name?: PropertyName | undefined;
     }
     /** Unlike ObjectLiteralElement, excludes JSXAttribute and JSXSpreadAttribute. */
     type ObjectLiteralElementLike = PropertyAssignment | ShorthandPropertyAssignment | SpreadAssignment | MethodDeclaration | AccessorDeclaration;
@@ -4605,8 +4605,8 @@
         readonly kind: SyntaxKind.ShorthandPropertyAssignment;
         readonly parent: ObjectLiteralExpression;
         readonly name: Identifier;
-        readonly equalsToken?: EqualsToken;
-        readonly objectAssignmentInitializer?: Expression;
+        readonly equalsToken?: EqualsToken | undefined;
+        readonly objectAssignmentInitializer?: Expression | undefined;
     }
     interface SpreadAssignment extends ObjectLiteralElement, JSDocContainer {
         readonly kind: SyntaxKind.SpreadAssignment;
@@ -4646,14 +4646,14 @@
     type FunctionLike = SignatureDeclaration;
     interface FunctionDeclaration extends FunctionLikeDeclarationBase, DeclarationStatement, LocalsContainer {
         readonly kind: SyntaxKind.FunctionDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
-        readonly name?: Identifier;
-        readonly body?: FunctionBody;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
+        readonly name?: Identifier | undefined;
+        readonly body?: FunctionBody | undefined;
     }
     interface MethodSignature extends SignatureDeclarationBase, TypeElement, LocalsContainer {
         readonly kind: SyntaxKind.MethodSignature;
         readonly parent: TypeLiteralNode | InterfaceDeclaration;
-        readonly modifiers?: NodeArray<Modifier>;
+        readonly modifiers?: NodeArray<Modifier> | undefined;
         readonly name: PropertyName;
     }
     interface MethodDeclaration extends FunctionLikeDeclarationBase, ClassElement, ObjectLiteralElement, JSDocContainer, LocalsContainer, FlowContainer {
@@ -4677,22 +4677,22 @@
     interface GetAccessorDeclaration extends FunctionLikeDeclarationBase, ClassElement, TypeElement, ObjectLiteralElement, JSDocContainer, LocalsContainer, FlowContainer {
         readonly kind: SyntaxKind.GetAccessor;
         readonly parent: ClassLikeDeclaration | ObjectLiteralExpression | TypeLiteralNode | InterfaceDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly name: PropertyName;
-        readonly body?: FunctionBody;
+        readonly body?: FunctionBody | undefined;
     }
     interface SetAccessorDeclaration extends FunctionLikeDeclarationBase, ClassElement, TypeElement, ObjectLiteralElement, JSDocContainer, LocalsContainer, FlowContainer {
         readonly kind: SyntaxKind.SetAccessor;
         readonly parent: ClassLikeDeclaration | ObjectLiteralExpression | TypeLiteralNode | InterfaceDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly name: PropertyName;
-        readonly body?: FunctionBody;
+        readonly body?: FunctionBody | undefined;
     }
     type AccessorDeclaration = GetAccessorDeclaration | SetAccessorDeclaration;
     interface IndexSignatureDeclaration extends SignatureDeclarationBase, ClassElement, TypeElement, LocalsContainer {
         readonly kind: SyntaxKind.IndexSignature;
         readonly parent: ObjectTypeDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly type: TypeNode;
     }
     interface ClassStaticBlockDeclaration extends ClassElement, JSDocContainer, LocalsContainer {
@@ -4711,15 +4711,15 @@
         readonly kind: SyntaxKind.ImportTypeAssertionContainer;
         readonly parent: ImportTypeNode;
         /** @deprecated */ readonly assertClause: AssertClause;
-        readonly multiLine?: boolean;
+        readonly multiLine?: boolean | undefined;
     }
     interface ImportTypeNode extends NodeWithTypeArguments {
         readonly kind: SyntaxKind.ImportType;
         readonly isTypeOf: boolean;
         readonly argument: TypeNode;
         /** @deprecated */ readonly assertions?: ImportTypeAssertionContainer;
-        readonly attributes?: ImportAttributes;
-        readonly qualifier?: EntityName;
+        readonly attributes?: ImportAttributes | undefined;
+        readonly qualifier?: EntityName | undefined;
     }
     interface ThisTypeNode extends TypeNode {
         readonly kind: SyntaxKind.ThisType;
@@ -4734,10 +4734,10 @@
     }
     interface ConstructorTypeNode extends FunctionOrConstructorTypeNodeBase, LocalsContainer {
         readonly kind: SyntaxKind.ConstructorType;
-        readonly modifiers?: NodeArray<Modifier>;
+        readonly modifiers?: NodeArray<Modifier> | undefined;
     }
     interface NodeWithTypeArguments extends TypeNode {
-        readonly typeArguments?: NodeArray<TypeNode>;
+        readonly typeArguments?: NodeArray<TypeNode> | undefined;
     }
     type TypeReferenceType = TypeReferenceNode | ExpressionWithTypeArguments;
     interface TypeReferenceNode extends NodeWithTypeArguments {
@@ -4747,9 +4747,9 @@
     interface TypePredicateNode extends TypeNode {
         readonly kind: SyntaxKind.TypePredicate;
         readonly parent: SignatureDeclaration | JSDocTypeExpression;
-        readonly assertsModifier?: AssertsKeyword;
+        readonly assertsModifier?: AssertsKeyword | undefined;
         readonly parameterName: Identifier | ThisTypeNode;
-        readonly type?: TypeNode;
+        readonly type?: TypeNode | undefined;
     }
     interface TypeQueryNode extends NodeWithTypeArguments {
         readonly kind: SyntaxKind.TypeQuery;
@@ -4769,9 +4769,9 @@
     }
     interface NamedTupleMember extends TypeNode, Declaration, JSDocContainer {
         readonly kind: SyntaxKind.NamedTupleMember;
-        readonly dotDotDotToken?: Token<SyntaxKind.DotDotDotToken>;
+        readonly dotDotDotToken?: Token<SyntaxKind.DotDotDotToken> | undefined;
         readonly name: Identifier;
-        readonly questionToken?: Token<SyntaxKind.QuestionToken>;
+        readonly questionToken?: Token<SyntaxKind.QuestionToken> | undefined;
         readonly type: TypeNode;
     }
     interface OptionalTypeNode extends TypeNode {
@@ -4818,13 +4818,13 @@
     }
     interface MappedTypeNode extends TypeNode, Declaration, LocalsContainer {
         readonly kind: SyntaxKind.MappedType;
-        readonly readonlyToken?: ReadonlyKeyword | PlusToken | MinusToken;
+        readonly readonlyToken?: ReadonlyKeyword | PlusToken | MinusToken | undefined;
         readonly typeParameter: TypeParameterDeclaration;
-        readonly nameType?: TypeNode;
-        readonly questionToken?: QuestionToken | PlusToken | MinusToken;
-        readonly type?: TypeNode;
+        readonly nameType?: TypeNode | undefined;
+        readonly questionToken?: QuestionToken | PlusToken | MinusToken | undefined;
+        readonly type?: TypeNode | undefined;
         /** Used only to produce grammar errors */
-        readonly members?: NodeArray<TypeElement>;
+        readonly members?: NodeArray<TypeElement> | undefined;
     }
     interface LiteralTypeNode extends TypeNode {
         readonly kind: SyntaxKind.LiteralType;
@@ -4922,14 +4922,14 @@
     }
     interface YieldExpression extends Expression {
         readonly kind: SyntaxKind.YieldExpression;
-        readonly asteriskToken?: AsteriskToken;
-        readonly expression?: Expression;
+        readonly asteriskToken?: AsteriskToken | undefined;
+        readonly expression?: Expression | undefined;
     }
     interface SyntheticExpression extends Expression {
         readonly kind: SyntaxKind.SyntheticExpression;
         readonly isSpread: boolean;
         readonly type: Type;
-        readonly tupleNameSource?: ParameterDeclaration | NamedTupleMember;
+        readonly tupleNameSource?: ParameterDeclaration | NamedTupleMember | undefined;
     }
     type ExponentiationOperator = SyntaxKind.AsteriskAsteriskToken;
     type MultiplicativeOperator = SyntaxKind.AsteriskToken | SyntaxKind.SlashToken | SyntaxKind.PercentToken;
@@ -4991,13 +4991,13 @@
     type ConciseBody = FunctionBody | Expression;
     interface FunctionExpression extends PrimaryExpression, FunctionLikeDeclarationBase, JSDocContainer, LocalsContainer, FlowContainer {
         readonly kind: SyntaxKind.FunctionExpression;
-        readonly modifiers?: NodeArray<Modifier>;
-        readonly name?: Identifier;
+        readonly modifiers?: NodeArray<Modifier> | undefined;
+        readonly name?: Identifier | undefined;
         readonly body: FunctionBody;
     }
     interface ArrowFunction extends Expression, FunctionLikeDeclarationBase, JSDocContainer, LocalsContainer, FlowContainer {
         readonly kind: SyntaxKind.ArrowFunction;
-        readonly modifiers?: NodeArray<Modifier>;
+        readonly modifiers?: NodeArray<Modifier> | undefined;
         readonly equalsGreaterThanToken: EqualsGreaterThanToken;
         readonly body: ConciseBody;
         readonly name: never;
@@ -5005,10 +5005,10 @@
     interface LiteralLikeNode extends Node {
         text: string;
         isUnterminated?: boolean;
-        hasExtendedUnicodeEscape?: boolean;
+        hasExtendedUnicodeEscape?: boolean | undefined;
     }
     interface TemplateLiteralLikeNode extends LiteralLikeNode {
-        rawText?: string;
+        rawText?: string | undefined;
     }
     interface LiteralExpression extends LiteralLikeNode, PrimaryExpression {
         _literalExpressionBrand: undefined;
@@ -5091,7 +5091,7 @@
     interface PropertyAccessExpression extends MemberExpression, NamedDeclaration, JSDocContainer, FlowContainer {
         readonly kind: SyntaxKind.PropertyAccessExpression;
         readonly expression: LeftHandSideExpression;
-        readonly questionDotToken?: QuestionDotToken;
+        readonly questionDotToken?: QuestionDotToken | undefined;
         readonly name: MemberName;
     }
     interface PropertyAccessChain extends PropertyAccessExpression {
@@ -5110,7 +5110,7 @@
     interface ElementAccessExpression extends MemberExpression, Declaration, JSDocContainer, FlowContainer {
         readonly kind: SyntaxKind.ElementAccessExpression;
         readonly expression: LeftHandSideExpression;
-        readonly questionDotToken?: QuestionDotToken;
+        readonly questionDotToken?: QuestionDotToken | undefined;
         readonly argumentExpression: Expression;
     }
     interface ElementAccessChain extends ElementAccessExpression {
@@ -5123,8 +5123,8 @@
     interface CallExpression extends LeftHandSideExpression, Declaration {
         readonly kind: SyntaxKind.CallExpression;
         readonly expression: LeftHandSideExpression;
-        readonly questionDotToken?: QuestionDotToken;
-        readonly typeArguments?: NodeArray<TypeNode>;
+        readonly questionDotToken?: QuestionDotToken | undefined;
+        readonly typeArguments?: NodeArray<TypeNode> | undefined;
         readonly arguments: NodeArray<Expression>;
     }
     interface CallChain extends CallExpression {
@@ -5144,13 +5144,13 @@
     interface NewExpression extends PrimaryExpression, Declaration {
         readonly kind: SyntaxKind.NewExpression;
         readonly expression: LeftHandSideExpression;
-        readonly typeArguments?: NodeArray<TypeNode>;
-        readonly arguments?: NodeArray<Expression>;
+        readonly typeArguments?: NodeArray<TypeNode> | undefined;
+        readonly arguments?: NodeArray<Expression> | undefined;
     }
     interface TaggedTemplateExpression extends MemberExpression {
         readonly kind: SyntaxKind.TaggedTemplateExpression;
         readonly tag: LeftHandSideExpression;
-        readonly typeArguments?: NodeArray<TypeNode>;
+        readonly typeArguments?: NodeArray<TypeNode> | undefined;
         readonly template: TemplateLiteral;
     }
     interface InstanceofExpression extends BinaryExpression {
@@ -5219,13 +5219,13 @@
         readonly kind: SyntaxKind.JsxOpeningElement;
         readonly parent: JsxElement;
         readonly tagName: JsxTagNameExpression;
-        readonly typeArguments?: NodeArray<TypeNode>;
+        readonly typeArguments?: NodeArray<TypeNode> | undefined;
         readonly attributes: JsxAttributes;
     }
     interface JsxSelfClosingElement extends PrimaryExpression {
         readonly kind: SyntaxKind.JsxSelfClosingElement;
         readonly tagName: JsxTagNameExpression;
-        readonly typeArguments?: NodeArray<TypeNode>;
+        readonly typeArguments?: NodeArray<TypeNode> | undefined;
         readonly attributes: JsxAttributes;
     }
     interface JsxFragment extends PrimaryExpression {
@@ -5246,7 +5246,7 @@
         readonly kind: SyntaxKind.JsxAttribute;
         readonly parent: JsxAttributes;
         readonly name: JsxAttributeName;
-        readonly initializer?: JsxAttributeValue;
+        readonly initializer?: JsxAttributeValue | undefined;
     }
     type JsxAttributeValue = StringLiteral | JsxExpression | JsxElement | JsxSelfClosingElement | JsxFragment;
     interface JsxSpreadAttribute extends ObjectLiteralElement {
@@ -5262,8 +5262,8 @@
     interface JsxExpression extends Expression {
         readonly kind: SyntaxKind.JsxExpression;
         readonly parent: JsxElement | JsxFragment | JsxAttributeLike;
-        readonly dotDotDotToken?: Token<SyntaxKind.DotDotDotToken>;
-        readonly expression?: Expression;
+        readonly dotDotDotToken?: Token<SyntaxKind.DotDotDotToken> | undefined;
+        readonly expression?: Expression | undefined;
     }
     interface JsxText extends LiteralLikeNode {
         readonly kind: SyntaxKind.JsxText;
@@ -5304,7 +5304,7 @@
     }
     interface VariableStatement extends Statement, FlowContainer {
         readonly kind: SyntaxKind.VariableStatement;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly declarationList: VariableDeclarationList;
     }
     interface ExpressionStatement extends Statement, FlowContainer {
@@ -5315,7 +5315,7 @@
         readonly kind: SyntaxKind.IfStatement;
         readonly expression: Expression;
         readonly thenStatement: Statement;
-        readonly elseStatement?: Statement;
+        readonly elseStatement?: Statement | undefined;
     }
     interface IterationStatement extends Statement {
         readonly statement: Statement;
@@ -5331,9 +5331,9 @@
     type ForInitializer = VariableDeclarationList | Expression;
     interface ForStatement extends IterationStatement, LocalsContainer, FlowContainer {
         readonly kind: SyntaxKind.ForStatement;
-        readonly initializer?: ForInitializer;
-        readonly condition?: Expression;
-        readonly incrementor?: Expression;
+        readonly initializer?: ForInitializer | undefined;
+        readonly condition?: Expression | undefined;
+        readonly incrementor?: Expression | undefined;
     }
     type ForInOrOfStatement = ForInStatement | ForOfStatement;
     interface ForInStatement extends IterationStatement, LocalsContainer, FlowContainer {
@@ -5343,22 +5343,22 @@
     }
     interface ForOfStatement extends IterationStatement, LocalsContainer, FlowContainer {
         readonly kind: SyntaxKind.ForOfStatement;
-        readonly awaitModifier?: AwaitKeyword;
+        readonly awaitModifier?: AwaitKeyword | undefined;
         readonly initializer: ForInitializer;
         readonly expression: Expression;
     }
     interface BreakStatement extends Statement, FlowContainer {
         readonly kind: SyntaxKind.BreakStatement;
-        readonly label?: Identifier;
+        readonly label?: Identifier | undefined;
     }
     interface ContinueStatement extends Statement, FlowContainer {
         readonly kind: SyntaxKind.ContinueStatement;
-        readonly label?: Identifier;
+        readonly label?: Identifier | undefined;
     }
     type BreakOrContinueStatement = BreakStatement | ContinueStatement;
     interface ReturnStatement extends Statement, FlowContainer {
         readonly kind: SyntaxKind.ReturnStatement;
-        readonly expression?: Expression;
+        readonly expression?: Expression | undefined;
     }
     interface WithStatement extends Statement, FlowContainer {
         readonly kind: SyntaxKind.WithStatement;
@@ -5400,13 +5400,13 @@
     interface TryStatement extends Statement, FlowContainer {
         readonly kind: SyntaxKind.TryStatement;
         readonly tryBlock: Block;
-        readonly catchClause?: CatchClause;
-        readonly finallyBlock?: Block;
+        readonly catchClause?: CatchClause | undefined;
+        readonly finallyBlock?: Block | undefined;
     }
     interface CatchClause extends Node, LocalsContainer {
         readonly kind: SyntaxKind.CatchClause;
         readonly parent: TryStatement;
-        readonly variableDeclaration?: VariableDeclaration;
+        readonly variableDeclaration?: VariableDeclaration | undefined;
         readonly block: Block;
     }
     type ObjectTypeDeclaration = ClassLikeDeclaration | InterfaceDeclaration | TypeLiteralNode;
@@ -5414,37 +5414,37 @@
     type DeclarationWithTypeParameterChildren = SignatureDeclaration | ClassLikeDeclaration | InterfaceDeclaration | TypeAliasDeclaration | JSDocTemplateTag;
     interface ClassLikeDeclarationBase extends NamedDeclaration, JSDocContainer {
         readonly kind: SyntaxKind.ClassDeclaration | SyntaxKind.ClassExpression;
-        readonly name?: Identifier;
-        readonly typeParameters?: NodeArray<TypeParameterDeclaration>;
-        readonly heritageClauses?: NodeArray<HeritageClause>;
+        readonly name?: Identifier | undefined;
+        readonly typeParameters?: NodeArray<TypeParameterDeclaration> | undefined;
+        readonly heritageClauses?: NodeArray<HeritageClause> | undefined;
         readonly members: NodeArray<ClassElement>;
     }
     interface ClassDeclaration extends ClassLikeDeclarationBase, DeclarationStatement {
         readonly kind: SyntaxKind.ClassDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         /** May be undefined in `export default class { ... }`. */
-        readonly name?: Identifier;
+        readonly name?: Identifier | undefined;
     }
     interface ClassExpression extends ClassLikeDeclarationBase, PrimaryExpression {
         readonly kind: SyntaxKind.ClassExpression;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
     }
     type ClassLikeDeclaration = ClassDeclaration | ClassExpression;
     interface ClassElement extends NamedDeclaration {
         _classElementBrand: undefined;
-        readonly name?: PropertyName;
+        readonly name?: PropertyName | undefined;
     }
     interface TypeElement extends NamedDeclaration {
         _typeElementBrand: undefined;
-        readonly name?: PropertyName;
+        readonly name?: PropertyName | undefined;
         readonly questionToken?: QuestionToken | undefined;
     }
     interface InterfaceDeclaration extends DeclarationStatement, JSDocContainer {
         readonly kind: SyntaxKind.InterfaceDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly name: Identifier;
-        readonly typeParameters?: NodeArray<TypeParameterDeclaration>;
-        readonly heritageClauses?: NodeArray<HeritageClause>;
+        readonly typeParameters?: NodeArray<TypeParameterDeclaration> | undefined;
+        readonly heritageClauses?: NodeArray<HeritageClause> | undefined;
         readonly members: NodeArray<TypeElement>;
     }
     interface HeritageClause extends Node {
@@ -5455,20 +5455,20 @@
     }
     interface TypeAliasDeclaration extends DeclarationStatement, JSDocContainer, LocalsContainer {
         readonly kind: SyntaxKind.TypeAliasDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly name: Identifier;
-        readonly typeParameters?: NodeArray<TypeParameterDeclaration>;
+        readonly typeParameters?: NodeArray<TypeParameterDeclaration> | undefined;
         readonly type: TypeNode;
     }
     interface EnumMember extends NamedDeclaration, JSDocContainer {
         readonly kind: SyntaxKind.EnumMember;
         readonly parent: EnumDeclaration;
         readonly name: PropertyName;
-        readonly initializer?: Expression;
+        readonly initializer?: Expression | undefined;
     }
     interface EnumDeclaration extends DeclarationStatement, JSDocContainer {
         readonly kind: SyntaxKind.EnumDeclaration;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly name: Identifier;
         readonly members: NodeArray<EnumMember>;
     }
@@ -5477,9 +5477,9 @@
     interface ModuleDeclaration extends DeclarationStatement, JSDocContainer, LocalsContainer {
         readonly kind: SyntaxKind.ModuleDeclaration;
         readonly parent: ModuleBody | SourceFile;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly name: ModuleName;
-        readonly body?: ModuleBody | JSDocNamespaceDeclaration;
+        readonly body?: ModuleBody | JSDocNamespaceDeclaration | undefined;
     }
     type NamespaceBody = ModuleBlock | NamespaceDeclaration;
     interface NamespaceDeclaration extends ModuleDeclaration {
@@ -5505,7 +5505,7 @@
     interface ImportEqualsDeclaration extends DeclarationStatement, JSDocContainer {
         readonly kind: SyntaxKind.ImportEqualsDeclaration;
         readonly parent: SourceFile | ModuleBlock;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly name: Identifier;
         readonly isTypeOnly: boolean;
         readonly moduleReference: ModuleReference;
@@ -5518,12 +5518,12 @@
     interface ImportDeclaration extends Statement {
         readonly kind: SyntaxKind.ImportDeclaration;
         readonly parent: SourceFile | ModuleBlock;
-        readonly modifiers?: NodeArray<ModifierLike>;
-        readonly importClause?: ImportClause;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
+        readonly importClause?: ImportClause | undefined;
         /** If this is not a StringLiteral it will be a grammar error. */
         readonly moduleSpecifier: Expression;
-        /** @deprecated */ readonly assertClause?: AssertClause;
-        readonly attributes?: ImportAttributes;
+        /** @deprecated */ readonly assertClause?: AssertClause | undefined;
+        readonly attributes?: ImportAttributes | undefined;
     }
     type NamedImportBindings = NamespaceImport | NamedImports;
     type NamedExportBindings = NamespaceExport | NamedExports;
@@ -5533,8 +5533,8 @@
         /** @deprecated Use `phaseModifier` instead */
         readonly isTypeOnly: boolean;
         readonly phaseModifier: undefined | ImportPhaseModifierSyntaxKind;
-        readonly name?: Identifier;
-        readonly namedBindings?: NamedImportBindings;
+        readonly name?: Identifier | undefined;
+        readonly namedBindings?: NamedImportBindings | undefined;
     }
     type ImportPhaseModifierSyntaxKind = SyntaxKind.TypeKeyword | SyntaxKind.DeferKeyword;
     /** @deprecated */
@@ -5557,7 +5557,7 @@
         readonly kind: SyntaxKind.ImportAttributes;
         readonly parent: ImportDeclaration | ExportDeclaration;
         readonly elements: NodeArray<ImportAttribute>;
-        readonly multiLine?: boolean;
+        readonly multiLine?: boolean | undefined;
     }
     interface NamespaceImport extends NamedDeclaration {
         readonly kind: SyntaxKind.NamespaceImport;
@@ -5576,14 +5576,14 @@
     interface ExportDeclaration extends DeclarationStatement, JSDocContainer {
         readonly kind: SyntaxKind.ExportDeclaration;
         readonly parent: SourceFile | ModuleBlock;
-        readonly modifiers?: NodeArray<ModifierLike>;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
         readonly isTypeOnly: boolean;
         /** Will not be assigned in the case of `export * from "foo";` */
-        readonly exportClause?: NamedExportBindings;
+        readonly exportClause?: NamedExportBindings | undefined;
         /** If this is not a StringLiteral it will be a grammar error. */
-        readonly moduleSpecifier?: Expression;
-        /** @deprecated */ readonly assertClause?: AssertClause;
-        readonly attributes?: ImportAttributes;
+        readonly moduleSpecifier?: Expression | undefined;
+        /** @deprecated */ readonly assertClause?: AssertClause | undefined;
+        readonly attributes?: ImportAttributes | undefined;
     }
     interface NamedImports extends Node {
         readonly kind: SyntaxKind.NamedImports;
@@ -5599,7 +5599,7 @@
     interface ImportSpecifier extends NamedDeclaration {
         readonly kind: SyntaxKind.ImportSpecifier;
         readonly parent: NamedImports;
-        readonly propertyName?: ModuleExportName;
+        readonly propertyName?: ModuleExportName | undefined;
         readonly name: Identifier;
         readonly isTypeOnly: boolean;
     }
@@ -5607,7 +5607,7 @@
         readonly kind: SyntaxKind.ExportSpecifier;
         readonly parent: NamedExports;
         readonly isTypeOnly: boolean;
-        readonly propertyName?: ModuleExportName;
+        readonly propertyName?: ModuleExportName | undefined;
         readonly name: ModuleExportName;
     }
     type ModuleExportName = Identifier | StringLiteral;
@@ -5665,8 +5665,8 @@
     interface ExportAssignment extends DeclarationStatement, JSDocContainer {
         readonly kind: SyntaxKind.ExportAssignment;
         readonly parent: SourceFile;
-        readonly modifiers?: NodeArray<ModifierLike>;
-        readonly isExportEquals?: boolean;
+        readonly modifiers?: NodeArray<ModifierLike> | undefined;
+        readonly isExportEquals?: boolean | undefined;
         readonly expression: Expression;
     }
     interface FileReference extends TextRange {
@@ -5679,7 +5679,7 @@
     }
     type CommentKind = SyntaxKind.SingleLineCommentTrivia | SyntaxKind.MultiLineCommentTrivia;
     interface CommentRange extends TextRange {
-        hasTrailingNewLine?: boolean;
+        hasTrailingNewLine?: boolean | undefined;
         kind: CommentKind;
     }
     interface SynthesizedComment extends CommentRange {
@@ -5740,27 +5740,27 @@
     interface JSDoc extends Node {
         readonly kind: SyntaxKind.JSDoc;
         readonly parent: HasJSDoc;
-        readonly tags?: NodeArray<JSDocTag>;
-        readonly comment?: string | NodeArray<JSDocComment>;
+        readonly tags?: NodeArray<JSDocTag> | undefined;
+        readonly comment?: string | NodeArray<JSDocComment> | undefined;
     }
     interface JSDocTag extends Node {
         readonly parent: JSDoc | JSDocTypeLiteral;
         readonly tagName: Identifier;
-        readonly comment?: string | NodeArray<JSDocComment>;
+        readonly comment?: string | NodeArray<JSDocComment> | undefined;
     }
     interface JSDocLink extends Node {
         readonly kind: SyntaxKind.JSDocLink;
-        readonly name?: EntityName | JSDocMemberName;
+        readonly name?: EntityName | JSDocMemberName | undefined;
         text: string;
     }
     interface JSDocLinkCode extends Node {
         readonly kind: SyntaxKind.JSDocLinkCode;
-        readonly name?: EntityName | JSDocMemberName;
+        readonly name?: EntityName | JSDocMemberName | undefined;
         text: string;
     }
     interface JSDocLinkPlain extends Node {
         readonly kind: SyntaxKind.JSDocLinkPlain;
-        readonly name?: EntityName | JSDocMemberName;
+        readonly name?: EntityName | JSDocMemberName | undefined;
         text: string;
     }
     type JSDocComment = JSDocText | JSDocLink | JSDocLinkCode | JSDocLinkPlain;
@@ -5827,7 +5827,7 @@
     }
     interface JSDocSeeTag extends JSDocTag {
         readonly kind: SyntaxKind.JSDocSeeTag;
-        readonly name?: JSDocNameReference;
+        readonly name?: JSDocNameReference | undefined;
     }
     interface JSDocReturnTag extends JSDocTag {
         readonly kind: SyntaxKind.JSDocReturnTag;
@@ -5840,15 +5840,15 @@
     interface JSDocTypedefTag extends JSDocTag, NamedDeclaration, LocalsContainer {
         readonly kind: SyntaxKind.JSDocTypedefTag;
         readonly parent: JSDoc;
-        readonly fullName?: JSDocNamespaceDeclaration | Identifier;
-        readonly name?: Identifier;
-        readonly typeExpression?: JSDocTypeExpression | JSDocTypeLiteral;
+        readonly fullName?: JSDocNamespaceDeclaration | Identifier | undefined;
+        readonly name?: Identifier | undefined;
+        readonly typeExpression?: JSDocTypeExpression | JSDocTypeLiteral | undefined;
     }
     interface JSDocCallbackTag extends JSDocTag, NamedDeclaration, LocalsContainer {
         readonly kind: SyntaxKind.JSDocCallbackTag;
         readonly parent: JSDoc;
-        readonly fullName?: JSDocNamespaceDeclaration | Identifier;
-        readonly name?: Identifier;
+        readonly fullName?: JSDocNamespaceDeclaration | Identifier | undefined;
+        readonly name?: Identifier | undefined;
         readonly typeExpression: JSDocSignature;
     }
     interface JSDocOverloadTag extends JSDocTag {
@@ -5862,14 +5862,14 @@
     }
     interface JSDocSignature extends JSDocType, Declaration, JSDocContainer, LocalsContainer {
         readonly kind: SyntaxKind.JSDocSignature;
-        readonly typeParameters?: readonly JSDocTemplateTag[];
+        readonly typeParameters?: readonly JSDocTemplateTag[] | undefined;
         readonly parameters: readonly JSDocParameterTag[];
         readonly type: JSDocReturnTag | undefined;
     }
     interface JSDocPropertyLikeTag extends JSDocTag, Declaration {
         readonly parent: JSDoc;
         readonly name: EntityName;
-        readonly typeExpression?: JSDocTypeExpression;
+        readonly typeExpression?: JSDocTypeExpression | undefined;
         /** Whether the property name came before the type -- non-standard for JSDoc, but Typescript-like */
         readonly isNameFirst: boolean;
         readonly isBracketed: boolean;
@@ -5882,7 +5882,7 @@
     }
     interface JSDocTypeLiteral extends JSDocType, Declaration {
         readonly kind: SyntaxKind.JSDocTypeLiteral;
-        readonly jsDocPropertyTags?: readonly JSDocPropertyLikeTag[];
+        readonly jsDocPropertyTags?: readonly JSDocPropertyLikeTag[] | undefined;
         /** If true, then this type literal represents an *array* of its type. */
         readonly isArrayType: boolean;
     }
@@ -5893,9 +5893,9 @@
     interface JSDocImportTag extends JSDocTag {
         readonly kind: SyntaxKind.JSDocImportTag;
         readonly parent: JSDoc;
-        readonly importClause?: ImportClause;
+        readonly importClause?: ImportClause | undefined;
         readonly moduleSpecifier: Expression;
-        readonly attributes?: ImportAttributes;
+        readonly attributes?: ImportAttributes | undefined;
     }
     type FlowType = Type | IncompleteType;
     interface IncompleteType {
@@ -5904,7 +5904,7 @@
     }
     interface AmdDependency {
         path: string;
-        name?: string;
+        name?: string | undefined;
     }
     /**
      * Subset of properties from SourceFile that are used in multiple utility functions
@@ -5964,6 +5964,7 @@
     }
     interface JsonSourceFile extends SourceFile {
         readonly statements: NodeArray<JsonObjectExpressionStatement>;
+        extendedSourceFiles?: string[];
     }
     interface TsConfigSourceFile extends JsonSourceFile {
         extendedSourceFiles?: string[];
@@ -6168,7 +6169,7 @@
         emitSkipped: boolean;
         /** Contains declaration emit diagnostics */
         diagnostics: readonly Diagnostic[];
-        emittedFiles?: string[];
+        emittedFiles?: string[] | undefined;
     }
     interface TypeChecker {
         getTypeOfSymbolAtLocation(symbol: Symbol, node: Node): Type;
@@ -6533,10 +6534,10 @@
     interface Symbol {
         flags: SymbolFlags;
         escapedName: __String;
-        declarations?: Declaration[];
-        valueDeclaration?: Declaration;
-        members?: SymbolTable;
-        exports?: SymbolTable;
+        declarations?: Declaration[] | undefined;
+        valueDeclaration?: Declaration | undefined;
+        members?: SymbolTable | undefined;
+        exports?: SymbolTable | undefined;
         globalExports?: SymbolTable;
     }
     interface Symbol {
@@ -6647,8 +6648,8 @@
         flags: TypeFlags;
         symbol: Symbol;
         pattern?: DestructuringPattern;
-        aliasSymbol?: Symbol;
-        aliasTypeArguments?: readonly Type[];
+        aliasSymbol?: Symbol | undefined;
+        aliasTypeArguments?: readonly Type[] | undefined;
     }
     interface Type {
         getFlags(): TypeFlags;
@@ -6819,11 +6820,11 @@
         checkType: Type;
         extendsType: Type;
         isDistributive: boolean;
-        inferTypeParameters?: TypeParameter[];
-        outerTypeParameters?: TypeParameter[];
-        instantiations?: Map<string, Type>;
-        aliasSymbol?: Symbol;
-        aliasTypeArguments?: Type[];
+        inferTypeParameters?: TypeParameter[] | undefined;
+        outerTypeParameters?: TypeParameter[] | undefined;
+        instantiations?: Map<string, Type> | undefined;
+        aliasSymbol?: Symbol | undefined;
+        aliasTypeArguments?: Type[] | undefined;
     }
     interface ConditionalType extends InstantiableType {
         root: ConditionalRoot;
@@ -6850,10 +6851,10 @@
         Construct = 1,
     }
     interface Signature {
-        declaration?: SignatureDeclaration | JSDocSignature;
-        typeParameters?: readonly TypeParameter[];
+        declaration?: SignatureDeclaration | JSDocSignature | undefined;
+        typeParameters?: readonly TypeParameter[] | undefined;
         parameters: readonly Symbol[];
-        thisParameter?: Symbol;
+        thisParameter?: Symbol | undefined;
     }
     interface Signature {
         getDeclaration(): SignatureDeclaration;
@@ -6875,8 +6876,8 @@
         keyType: Type;
         type: Type;
         isReadonly: boolean;
-        declaration?: IndexSignatureDeclaration;
-        components?: ElementWithComputedPropertyName[];
+        declaration?: IndexSignatureDeclaration | undefined;
+        components?: ElementWithComputedPropertyName[] | undefined;
     }
     enum InferencePriority {
         None = 0,
@@ -6905,8 +6906,8 @@
         category: DiagnosticCategory;
         code: number;
         message: string;
-        reportsUnnecessary?: {};
-        reportsDeprecated?: {};
+        reportsUnnecessary?: {} | undefined;
+        reportsDeprecated?: {} | undefined;
     }
     /**
      * A linked list of formatted diagnostic messages to be used as part of a multiline message.
@@ -6918,14 +6919,14 @@
         messageText: string;
         category: DiagnosticCategory;
         code: number;
-        next?: DiagnosticMessageChain[];
+        next?: DiagnosticMessageChain[] | undefined;
     }
     interface Diagnostic extends DiagnosticRelatedInformation {
         /** May store more in future. For now, this will simply be `true` to indicate when a diagnostic is an unused-identifier diagnostic. */
-        reportsUnnecessary?: {};
-        reportsDeprecated?: {};
-        source?: string;
-        relatedInformation?: DiagnosticRelatedInformation[];
+        reportsUnnecessary?: {} | undefined;
+        reportsDeprecated?: {} | undefined;
+        source?: string | undefined;
+        relatedInformation?: DiagnosticRelatedInformation[] | undefined;
     }
     interface DiagnosticRelatedInformation {
         category: DiagnosticCategory;
@@ -6986,9 +6987,9 @@
         /** The path as the user originally wrote it */
         originalPath?: string;
         /** @deprecated */
-        prepend?: boolean;
+        prepend?: boolean | undefined;
         /** True if it is intended that this reference form a circularity */
-        circular?: boolean;
+        circular?: boolean | undefined;
     }
     enum WatchFileKind {
         FixedPollingInterval = 0,
@@ -7057,7 +7058,7 @@
         locale?: string;
         mapRoot?: string;
         maxNodeModuleJsDepth?: number;
-        module?: ModuleKind;
+        module?: ModuleKind | undefined;
         moduleResolution?: ModuleResolutionKind;
         moduleSuffixes?: string[];
         moduleDetection?: ModuleDetectionKind;
@@ -7121,8 +7122,8 @@
         suppressExcessPropertyErrors?: boolean;
         /** @deprecated */
         suppressImplicitAnyIndexErrors?: boolean;
-        target?: ScriptTarget;
-        traceResolution?: boolean;
+        target?: ScriptTarget | undefined;
+        traceResolution?: boolean | undefined;
         useUnknownInCatchVariables?: boolean;
         noUncheckedSideEffectImports?: boolean;
         resolveJsonModule?: boolean;
@@ -7138,7 +7139,7 @@
     interface WatchOptions {
         watchFile?: WatchFileKind;
         watchDirectory?: WatchDirectoryKind;
-        fallbackPolling?: PollingWatchKind;
+        fallbackPolling?: PollingWatchKind | undefined;
         synchronousWatchDirectory?: boolean;
         excludeDirectories?: string[];
         excludeFiles?: string[];
@@ -7238,8 +7239,8 @@
         options: CompilerOptions;
         typeAcquisition?: TypeAcquisition;
         fileNames: string[];
-        projectReferences?: readonly ProjectReference[];
-        watchOptions?: WatchOptions;
+        projectReferences?: readonly ProjectReference[] | undefined;
+        watchOptions?: WatchOptions | undefined;
         raw?: any;
         errors: Diagnostic[];
         wildcardDirectories?: MapLike<WatchDirectoryFlags>;
@@ -7252,23 +7253,39 @@
     interface CreateProgramOptions {
         rootNames: readonly string[];
         options: CompilerOptions;
-        projectReferences?: readonly ProjectReference[];
-        host?: CompilerHost;
-        oldProgram?: Program;
-        configFileParsingDiagnostics?: readonly Diagnostic[];
+        projectReferences?: readonly ProjectReference[] | undefined;
+        host?: CompilerHost | undefined;
+        oldProgram?: Program | undefined;
+        configFileParsingDiagnostics?: readonly Diagnostic[] | undefined;
     }
     interface ModuleResolutionHost {
         fileExists(fileName: string): boolean;
         readFile(fileName: string): string | undefined;
-        trace?(s: string): void;
-        directoryExists?(directoryName: string): boolean;
+        trace?:
+            | {
+                trace(s: string): void;
+            }["trace"]
+            | undefined;
+        directoryExists?:
+            | {
+                directoryExists(directoryName: string): boolean;
+            }["directoryExists"]
+            | undefined;
         /**
          * Resolve a symbolic link.
          * @see https://nodejs.org/api/fs.html#fs_fs_realpathsync_path_options
          */
-        realpath?(path: string): string;
+        realpath?:
+            | {
+                realpath(path: string): string;
+            }["realpath"]
+            | undefined;
         getCurrentDirectory?(): string;
-        getDirectories?(path: string): string[];
+        getDirectories?:
+            | {
+                getDirectories(path: string): string[];
+            }["getDirectories"]
+            | undefined;
         useCaseSensitiveFileNames?: boolean | (() => boolean) | undefined;
     }
     /**
@@ -7361,13 +7378,21 @@
         getSourceFileByPath?(fileName: string, path: Path, languageVersionOrOptions: ScriptTarget | CreateSourceFileOptions, onError?: (message: string) => void, shouldCreateNewSourceFile?: boolean): SourceFile | undefined;
         getCancellationToken?(): CancellationToken;
         getDefaultLibFileName(options: CompilerOptions): string;
-        getDefaultLibLocation?(): string;
+        getDefaultLibLocation?:
+            | {
+                getDefaultLibLocation(): string;
+            }["getDefaultLibLocation"]
+            | undefined;
         writeFile: WriteFileCallback;
         getCurrentDirectory(): string;
         getCanonicalFileName(fileName: string): string;
         useCaseSensitiveFileNames(): boolean;
         getNewLine(): string;
-        readDirectory?(rootDir: string, extensions: readonly string[], excludes: readonly string[] | undefined, includes: readonly string[], depth?: number): string[];
+        readDirectory?:
+            | {
+                readDirectory(rootDir: string, extensions: readonly string[], excludes: readonly string[] | undefined, includes: readonly string[], depth?: number): string[];
+            }["readDirectory"]
+            | undefined;
         /** @deprecated supply resolveModuleNameLiterals instead for resolution that can handle newer resolution modes like nodenext */
         resolveModuleNames?(moduleNames: string[], containingFile: string, reusedNames: string[] | undefined, redirectedReference: ResolvedProjectReference | undefined, options: CompilerOptions, containingSourceFile?: SourceFile): (ResolvedModule | undefined)[];
         /**
@@ -7385,9 +7410,13 @@
         getEnvironmentVariable?(name: string): string | undefined;
         /** If provided along with custom resolveModuleNames or resolveTypeReferenceDirectives, used to determine if unchanged file path needs to re-resolve modules/type reference directives */
         hasInvalidatedResolutions?(filePath: Path): boolean;
-        createHash?(data: string): string;
+        createHash?:
+            | {
+                createHash(data: string): string;
+            }["createHash"]
+            | undefined;
         getParsedCommandLine?(fileName: string): ParsedCommandLine | undefined;
-        jsDocParsingMode?: JSDocParsingMode;
+        jsDocParsingMode?: JSDocParsingMode | undefined;
     }
     interface SourceMapRange extends TextRange {
         source?: SourceMapSource;
@@ -8177,10 +8206,10 @@
         substituteNode?(hint: EmitHint, node: Node): Node;
     }
     interface PrinterOptions {
-        removeComments?: boolean;
-        newLine?: NewLineKind;
+        removeComments?: boolean | undefined;
+        newLine?: NewLineKind | undefined;
         omitTrailingSemicolon?: boolean;
-        noEmitHelpers?: boolean;
+        noEmitHelpers?: boolean | undefined;
     }
     interface GetEffectiveTypeRootsHost {
         getCurrentDirectory?(): string;
@@ -8344,7 +8373,7 @@
         readonly allowIncompleteCompletions?: boolean;
         readonly importModuleSpecifierPreference?: "shortest" | "project-relative" | "relative" | "non-relative";
         /** Determines whether we import `foo/index.ts` as "foo", "foo/index", or "foo/index.js" */
-        readonly importModuleSpecifierEnding?: "auto" | "minimal" | "index" | "js";
+        readonly importModuleSpecifierEnding?: "auto" | "minimal" | "index" | "js" | undefined;
         readonly allowTextChangesInNewFiles?: boolean;
         readonly providePrefixAndSuffixTextForRename?: boolean;
         readonly includePackageJsonAutoImports?: "auto" | "on" | "off";
@@ -9182,7 +9211,7 @@
     function isQuestionOrPlusOrMinusToken(node: Node): node is QuestionToken | PlusToken | MinusToken;
     function isModuleName(node: Node): node is ModuleName;
     function isBinaryOperatorToken(node: Node): node is BinaryOperatorToken;
-    function setTextRange<T extends TextRange>(range: T, location: TextRange | undefined): T;
+    function setTextRange<T extends TextRange>(range: T, location: Readonly<TextRange> | undefined): T;
     function canHaveModifiers(node: Node): node is HasModifiers;
     function canHaveDecorators(node: Node): node is HasDecorators;
     /**
@@ -9223,7 +9252,7 @@
          * check specified by `isFileProbablyExternalModule` will be used to set the field.
          */
         setExternalModuleIndicator?: (file: SourceFile) => void;
-        jsDocParsingMode?: JSDocParsingMode;
+        jsDocParsingMode?: JSDocParsingMode | undefined;
     }
     function parseCommandLine(commandLine: readonly string[], readFile?: (path: string) => string | undefined): ParsedCommandLine;
     function parseBuildCommand(commandLine: readonly string[]): ParsedBuildCommand;
@@ -9237,7 +9266,7 @@
      */
     function readConfigFile(fileName: string, readFile: (path: string) => string | undefined): {
         config?: any;
-        error?: Diagnostic;
+        error?: Diagnostic | undefined;
     };
     /**
      * Parse the text of the tsconfig.json file
@@ -9246,7 +9275,7 @@
      */
     function parseConfigFileTextToJson(fileName: string, jsonText: string): {
         config?: any;
-        error?: Diagnostic;
+        error?: Diagnostic | undefined;
     };
     /**
      * Read tsconfig.json file
@@ -9307,12 +9336,12 @@
     interface ParsedTsconfig {
         raw: any;
         options?: CompilerOptions;
-        watchOptions?: WatchOptions;
+        watchOptions?: WatchOptions | undefined;
         typeAcquisition?: TypeAcquisition;
         /**
          * Note that the case of the config path has not yet been normalized, as no files have been imported into the project yet
          */
-        extendedConfigPath?: string | string[];
+        extendedConfigPath?: string | string[] | undefined;
     }
     interface ExtendedConfigCacheEntry {
         extendedResult: TsConfigSourceFile;
@@ -9671,7 +9700,7 @@
         /**
          * If provided this would be used this hash instead of actual file shape text for detecting changes
          */
-        createHash?: (data: string) => string;
+        createHash?: ((data: string) => string) | undefined;
         /**
          * When emit or emitNextAffectedFile are called without writeFile,
          * this callback if present would be used to write files
@@ -9795,7 +9824,7 @@
         rootNames: readonly string[];
         options: CompilerOptions;
         configFileParsingDiagnostics?: readonly Diagnostic[];
-        projectReferences?: readonly ProjectReference[];
+        projectReferences?: readonly ProjectReference[] | undefined;
         host?: CompilerHost;
         createProgram?: CreateProgram<T>;
     }
@@ -9814,7 +9843,7 @@
         setTimeout?(callback: (...args: any[]) => void, ms: number, ...args: any[]): any;
         /** If provided, will be used to reset existing delayed compilation */
         clearTimeout?(timeoutId: any): void;
-        preferNonRecursiveWatch?: boolean;
+        preferNonRecursiveWatch?: boolean | undefined;
     }
     interface ProgramHost<T extends BuilderProgram> {
         /**
@@ -9826,7 +9855,11 @@
         getCurrentDirectory(): string;
         getDefaultLibFileName(options: CompilerOptions): string;
         getDefaultLibLocation?(): string;
-        createHash?(data: string): string;
+        createHash?:
+            | {
+                createHash(data: string): string;
+            }["createHash"]
+            | undefined;
         /**
          * Use to check file presence for source files and
          * if resolveModuleNames is not provided (complier is in charge of module resolution) then module files as well
@@ -9844,11 +9877,19 @@
         /** If provided, used to cache and handle directory structure modifications */
         readDirectory?(path: string, extensions?: readonly string[], exclude?: readonly string[], include?: readonly string[], depth?: number): string[];
         /** Symbol links resolution */
-        realpath?(path: string): string;
+        realpath?:
+            | {
+                realpath(path: string): string;
+            }["realpath"]
+            | undefined;
         /** If provided would be used to write log about compilation */
         trace?(s: string): void;
         /** If provided is used to get the environment variable */
-        getEnvironmentVariable?(name: string): string | undefined;
+        getEnvironmentVariable?:
+            | {
+                getEnvironmentVariable(name: string): string | undefined;
+            }["getEnvironmentVariable"]
+            | undefined;
         /**
          * @deprecated supply resolveModuleNameLiterals instead for resolution that can handle newer resolution modes like nodenext
          *
@@ -9887,9 +9928,9 @@
         rootFiles: string[];
         /** Compiler options */
         options: CompilerOptions;
-        watchOptions?: WatchOptions;
+        watchOptions?: WatchOptions | undefined;
         /** Project References */
-        projectReferences?: readonly ProjectReference[];
+        projectReferences?: readonly ProjectReference[] | undefined;
     }
     /**
      * Host to create watch with config file
@@ -9898,9 +9939,9 @@
         /** Name of the config file to compile */
         configFileName: string;
         /** Options to extend */
-        optionsToExtend?: CompilerOptions;
-        watchOptionsToExtend?: WatchOptions;
-        extraFileExtensions?: readonly FileExtensionInfo[];
+        optionsToExtend?: CompilerOptions | undefined;
+        watchOptionsToExtend?: WatchOptions | undefined;
+        extraFileExtensions?: readonly FileExtensionInfo[] | undefined;
         /**
          * Used to generate source file names from the config file and its include, exclude, files rules
          * and also to cache the directory stucture
@@ -9970,7 +10011,7 @@
         afterProgramEmitAndDiagnostics?(program: T): void;
     }
     interface SolutionBuilderHost<T extends BuilderProgram> extends SolutionBuilderHostBase<T> {
-        reportErrorSummary?: ReportEmitErrorSummary;
+        reportErrorSummary?: ReportEmitErrorSummary | undefined;
     }
     interface SolutionBuilderWithWatchHost<T extends BuilderProgram> extends SolutionBuilderHostBase<T>, WatchHost {
     }
