--- a/stage1/cohere/markdownblocks/decodeString.ts
+++ b/stage1/cohere/markdownblocks/decodeString.ts
@@ -47,7 +47,7 @@
         (code > 13 && code < 32) ||
         (code > 126 && code < 160) ||
         (code > 55295 && code < 57344) ||
-        (code > 64975 && code < 65008) ||
+        (code > 64974 && code < 65008) ||
         (code & 65535) === 65535 ||
         (code & 65535) === 65534 ||
         code > 1114111
