--- 51aa3a96
+++ 6a056842
@@ -79,8 +79,8 @@
 static const char *const adamic_shape_0_names[] = {"label", "#secret@1"};
 static const bool adamic_shape_0_references[] = {true, true};
 static const adamic_shape adamic_shape_0 = {2, adamic_shape_0_names, adamic_shape_0_references, &adamic_shape_0_methods};
+static adamic_slot_cache adamic_cache_15;
 static adamic_slot_cache adamic_cache_16;
-static adamic_slot_cache adamic_cache_17;
 static adamic_method_function adamic_method_10;
 static adamic_value adamic_method_10(adamic_object *self, adamic_value *arguments) {
 	(void)arguments;
@@ -108,63 +108,63 @@
 static const char *const adamic_shape_1_names[] = {"label", "#secret@1", "child", "#parent@2", "#secret@2"};
 static const bool adamic_shape_1_references[] = {true, true, true, true, true};
 static const adamic_shape adamic_shape_1 = {5, adamic_shape_1_names, adamic_shape_1_references, &adamic_shape_1_methods};
+static adamic_slot_cache adamic_cache_19;
 static adamic_slot_cache adamic_cache_20;
-static adamic_slot_cache adamic_cache_21;
+static adamic_slot_cache adamic_cache_23;
 static adamic_slot_cache adamic_cache_24;
-static adamic_slot_cache adamic_cache_25;
+static adamic_slot_cache adamic_cache_28;
 static adamic_slot_cache adamic_cache_29;
-static adamic_slot_cache adamic_cache_30;
+static adamic_slot_cache adamic_cache_33;
 static adamic_slot_cache adamic_cache_34;
-static adamic_slot_cache adamic_cache_35;
+static adamic_slot_cache adamic_cache_44;
 static adamic_slot_cache adamic_cache_45;
-static adamic_slot_cache adamic_cache_46;
+static adamic_slot_cache adamic_cache_60;
 static adamic_slot_cache adamic_cache_61;
 static adamic_slot_cache adamic_cache_62;
-static adamic_slot_cache adamic_cache_63;
+static adamic_slot_cache adamic_cache_66;
 static adamic_slot_cache adamic_cache_67;
-static adamic_slot_cache adamic_cache_68;
+static adamic_slot_cache adamic_cache_72;
 static adamic_slot_cache adamic_cache_73;
-static adamic_slot_cache adamic_cache_74;
-static adamic_slot_cache adamic_cache_76;
-static adamic_slot_cache adamic_cache_80;
-static adamic_slot_cache adamic_cache_85;
+static adamic_slot_cache adamic_cache_75;
+static adamic_slot_cache adamic_cache_79;
+static adamic_slot_cache adamic_cache_84;
 static const char *const adamic_shape_2_names[] = {"value", "contents"};
 static const bool adamic_shape_2_references[] = {true, true};
 static const adamic_shape adamic_shape_2 = {2, adamic_shape_2_names, adamic_shape_2_references, NULL};
+static adamic_slot_cache adamic_cache_94;
 static adamic_slot_cache adamic_cache_95;
 static adamic_slot_cache adamic_cache_96;
-static adamic_slot_cache adamic_cache_97;
 static const char *const adamic_shape_3_names[] = {"text"};
 static const bool adamic_shape_3_references[] = {true};
 static const adamic_shape adamic_shape_3 = {1, adamic_shape_3_names, adamic_shape_3_references, NULL};
+static adamic_slot_cache adamic_cache_102;
 static adamic_slot_cache adamic_cache_103;
-static adamic_slot_cache adamic_cache_104;
+static adamic_slot_cache adamic_cache_108;
 static adamic_slot_cache adamic_cache_109;
 static adamic_slot_cache adamic_cache_110;
-static adamic_slot_cache adamic_cache_111;
+static adamic_slot_cache adamic_cache_115;
 static adamic_slot_cache adamic_cache_116;
 static adamic_slot_cache adamic_cache_117;
-static adamic_slot_cache adamic_cache_118;
+static adamic_slot_cache adamic_cache_122;
 static adamic_slot_cache adamic_cache_123;
 static adamic_slot_cache adamic_cache_124;
-static adamic_slot_cache adamic_cache_125;
+static adamic_slot_cache adamic_cache_130;
 static adamic_slot_cache adamic_cache_131;
-static adamic_slot_cache adamic_cache_132;
+static adamic_slot_cache adamic_cache_136;
 static adamic_slot_cache adamic_cache_137;
 static adamic_slot_cache adamic_cache_138;
-static adamic_slot_cache adamic_cache_139;
+static adamic_slot_cache adamic_cache_143;
 static adamic_slot_cache adamic_cache_144;
 static adamic_slot_cache adamic_cache_145;
-static adamic_slot_cache adamic_cache_146;
-static adamic_slot_cache adamic_cache_178;
+static adamic_slot_cache adamic_cache_174;
 static const char *const adamic_shape_4_names[] = {"#visible", "shown"};
 static const bool adamic_shape_4_references[] = {true, true};
 static const adamic_shape adamic_shape_4 = {2, adamic_shape_4_names, adamic_shape_4_references, NULL};
-static adamic_slot_cache adamic_cache_200;
-static adamic_slot_cache adamic_cache_201;
-static adamic_slot_cache adamic_cache_202;
-static adamic_slot_cache adamic_cache_205;
-static adamic_slot_cache adamic_cache_209;
+static adamic_slot_cache adamic_cache_194;
+static adamic_slot_cache adamic_cache_195;
+static adamic_slot_cache adamic_cache_196;
+static adamic_slot_cache adamic_cache_199;
+static adamic_slot_cache adamic_cache_203;
 static const char *const adamic_shape_5_names[] = {"label"};
 static const bool adamic_shape_5_references[] = {true};
 static const adamic_shape adamic_shape_5 = {1, adamic_shape_5_names, adamic_shape_5_references, NULL};
@@ -184,73 +184,71 @@
 	ADAMIC_CHECK_STACK();
 	adamic_object * adamic_temporary_1 = adamic_function_1_Vault_new(adamic_local_5_value);
 	adamic_object * adamic_local_6_source = adamic_temporary_1;
-	adamic_object * adamic_temporary_2 = adamic_function_6_checked_view_members(adamic_local_6_source);
-	adamic_object * adamic_temporary_3 = adamic_retain(adamic_temporary_2);
-	adamic_object * adamic_temporary_4 = adamic_object_copy(adamic_temporary_3);
-	adamic_object * adamic_temporary_5 = adamic_temporary_4;
-	adamic_release(adamic_temporary_3);
+	adamic_object * adamic_temporary_2 = adamic_retain(adamic_local_6_source);
+	adamic_object * adamic_temporary_3 = adamic_object_copy(adamic_temporary_2);
+	adamic_object * adamic_temporary_4 = adamic_temporary_3;
 	adamic_release(adamic_temporary_2);
 	adamic_release(adamic_local_6_source);
-	return adamic_temporary_5;
+	return adamic_temporary_4;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_1_Vault_new(adamic_string * adamic_local_14_value) {
 	ADAMIC_CHECK_STACK();
-	adamic_object * adamic_temporary_6 = adamic_object_new(&adamic_shape_0);
-	adamic_temporary_6->class = &adamic_class_1;
-	adamic_temporary_6->slots[0].reference = NULL;
-	adamic_temporary_6->slots[1].reference = NULL;
-	adamic_object * adamic_local_13_this = adamic_temporary_6;
+	adamic_object * adamic_temporary_5 = adamic_object_new(&adamic_shape_0);
+	adamic_temporary_5->class = &adamic_class_1;
+	adamic_temporary_5->slots[0].reference = NULL;
+	adamic_temporary_5->slots[1].reference = NULL;
+	adamic_object * adamic_local_13_this = adamic_temporary_5;
 	adamic_function_5_Vault_new_initialize(adamic_local_13_this, adamic_local_14_value);
-	adamic_object * adamic_temporary_7 = adamic_retain(adamic_local_13_this);
+	adamic_object * adamic_temporary_6 = adamic_retain(adamic_local_13_this);
 	adamic_release(adamic_local_13_this);
-	return adamic_temporary_7;
+	return adamic_temporary_6;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_1_Vault_new_in(adamic_region *region, adamic_string * adamic_local_14_value) {
 	ADAMIC_CHECK_STACK();
-	adamic_object *adamic_temporary_8 = adamic_object_new_in(region, &adamic_shape_0);
-	adamic_temporary_8->class = &adamic_class_1;
-	adamic_temporary_8->slots[0].reference = NULL;
-	adamic_temporary_8->slots[1].reference = NULL;
-	adamic_object * adamic_local_13_this = adamic_retain(adamic_temporary_8);
+	adamic_object *adamic_temporary_7 = adamic_object_new_in(region, &adamic_shape_0);
+	adamic_temporary_7->class = &adamic_class_1;
+	adamic_temporary_7->slots[0].reference = NULL;
+	adamic_temporary_7->slots[1].reference = NULL;
+	adamic_object * adamic_local_13_this = adamic_retain(adamic_temporary_7);
 	adamic_function_5_Vault_new_initialize(adamic_local_13_this, adamic_local_14_value);
-	adamic_object * adamic_temporary_9 = adamic_local_13_this;
+	adamic_object * adamic_temporary_8 = adamic_local_13_this;
 	adamic_release(adamic_local_13_this);
-	return adamic_temporary_9;
+	return adamic_temporary_8;
 	adamic_unreachable();
 }
 
 static adamic_string * adamic_function_2_Vault_read1(adamic_object * adamic_local_7_this) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_10 = ((adamic_string *)(&adamic_local_7_this->slots[1])->reference);
-	adamic_string * adamic_temporary_11 = adamic_retain(adamic_temporary_10);
-	return adamic_temporary_11;
+	adamic_string * adamic_temporary_9 = ((adamic_string *)(&adamic_local_7_this->slots[1])->reference);
+	adamic_string * adamic_temporary_10 = adamic_retain(adamic_temporary_9);
+	return adamic_temporary_10;
 	adamic_unreachable();
 }
 
 static adamic_string * adamic_function_3_Vault_read(adamic_object * adamic_local_8_this) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_12 = adamic_function_2_Vault_read1(adamic_local_8_this);
-	adamic_string * adamic_temporary_13 = adamic_temporary_12;
-	return adamic_temporary_13;
+	adamic_string * adamic_temporary_11 = adamic_function_2_Vault_read1(adamic_local_8_this);
+	adamic_string * adamic_temporary_12 = adamic_temporary_11;
+	return adamic_temporary_12;
 	adamic_unreachable();
 }
 
 static void adamic_function_4_Vault_replace(adamic_object * adamic_local_9_this, adamic_string * adamic_local_10_value) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_14 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_2, adamic_local_10_value});
+	adamic_string * adamic_temporary_13 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_2, adamic_local_10_value});
 	if (adamic_local_9_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting '#secret@1')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_9_this, "#secret@1");
-	adamic_value *adamic_temporary_15 = (adamic_local_9_this->shape == &adamic_shape_0 || adamic_local_9_this->shape == &adamic_shape_1 ? (&adamic_local_9_this->slots[1]) : adamic_object_data_field(adamic_local_9_this, "#secret@1", &adamic_cache_17));
-	void *adamic_temporary_18 = adamic_temporary_15->reference;
-	adamic_temporary_15->reference = adamic_temporary_14;
-	if (adamic_temporary_18 != NULL) adamic_release(adamic_temporary_18);
+	adamic_value *adamic_temporary_14 = (adamic_local_9_this->shape == &adamic_shape_0 || adamic_local_9_this->shape == &adamic_shape_1 ? (&adamic_local_9_this->slots[1]) : adamic_object_data_field(adamic_local_9_this, "#secret@1", &adamic_cache_16));
+	void *adamic_temporary_17 = adamic_temporary_14->reference;
+	adamic_temporary_14->reference = adamic_temporary_13;
+	if (adamic_temporary_17 != NULL) adamic_release(adamic_temporary_17);
 	return;
 }
 
@@ -261,142 +259,142 @@
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_11_this, "label");
-	adamic_value *adamic_temporary_19 = (adamic_local_11_this->shape == &adamic_shape_0 || adamic_local_11_this->shape == &adamic_shape_1 ? (&adamic_local_11_this->slots[0]) : adamic_object_data_field(adamic_local_11_this, "label", &adamic_cache_21));
-	void *adamic_temporary_22 = adamic_temporary_19->reference;
-	adamic_temporary_19->reference = NULL;
-	if (adamic_temporary_22 != NULL) adamic_release(adamic_temporary_22);
+	adamic_value *adamic_temporary_18 = (adamic_local_11_this->shape == &adamic_shape_0 || adamic_local_11_this->shape == &adamic_shape_1 ? (&adamic_local_11_this->slots[0]) : adamic_object_data_field(adamic_local_11_this, "label", &adamic_cache_20));
+	void *adamic_temporary_21 = adamic_temporary_18->reference;
+	adamic_temporary_18->reference = NULL;
+	if (adamic_temporary_21 != NULL) adamic_release(adamic_temporary_21);
 	if (adamic_local_11_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting '#secret@1')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_11_this, "#secret@1");
-	adamic_value *adamic_temporary_23 = (adamic_local_11_this->shape == &adamic_shape_0 || adamic_local_11_this->shape == &adamic_shape_1 ? (&adamic_local_11_this->slots[1]) : adamic_object_data_field(adamic_local_11_this, "#secret@1", &adamic_cache_25));
-	void *adamic_temporary_26 = adamic_temporary_23->reference;
-	adamic_temporary_23->reference = NULL;
-	if (adamic_temporary_26 != NULL) adamic_release(adamic_temporary_26);
-	adamic_string * adamic_temporary_27 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_0, adamic_local_12_value});
+	adamic_value *adamic_temporary_22 = (adamic_local_11_this->shape == &adamic_shape_0 || adamic_local_11_this->shape == &adamic_shape_1 ? (&adamic_local_11_this->slots[1]) : adamic_object_data_field(adamic_local_11_this, "#secret@1", &adamic_cache_24));
+	void *adamic_temporary_25 = adamic_temporary_22->reference;
+	adamic_temporary_22->reference = NULL;
+	if (adamic_temporary_25 != NULL) adamic_release(adamic_temporary_25);
+	adamic_string * adamic_temporary_26 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_0, adamic_local_12_value});
 	if (adamic_local_11_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'label')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_11_this, "label");
-	adamic_value *adamic_temporary_28 = (adamic_local_11_this->shape == &adamic_shape_0 || adamic_local_11_this->shape == &adamic_shape_1 ? (&adamic_local_11_this->slots[0]) : adamic_object_data_field(adamic_local_11_this, "label", &adamic_cache_30));
-	void *adamic_temporary_31 = adamic_temporary_28->reference;
-	adamic_temporary_28->reference = adamic_temporary_27;
-	if (adamic_temporary_31 != NULL) adamic_release(adamic_temporary_31);
-	adamic_string * adamic_temporary_32 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_1, adamic_local_12_value});
+	adamic_value *adamic_temporary_27 = (adamic_local_11_this->shape == &adamic_shape_0 || adamic_local_11_this->shape == &adamic_shape_1 ? (&adamic_local_11_this->slots[0]) : adamic_object_data_field(adamic_local_11_this, "label", &adamic_cache_29));
+	void *adamic_temporary_30 = adamic_temporary_27->reference;
+	adamic_temporary_27->reference = adamic_temporary_26;
+	if (adamic_temporary_30 != NULL) adamic_release(adamic_temporary_30);
+	adamic_string * adamic_temporary_31 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_1, adamic_local_12_value});
 	if (adamic_local_11_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting '#secret@1')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_11_this, "#secret@1");
-	adamic_value *adamic_temporary_33 = (adamic_local_11_this->shape == &adamic_shape_0 || adamic_local_11_this->shape == &adamic_shape_1 ? (&adamic_local_11_this->slots[1]) : adamic_object_data_field(adamic_local_11_this, "#secret@1", &adamic_cache_35));
-	void *adamic_temporary_36 = adamic_temporary_33->reference;
-	adamic_temporary_33->reference = adamic_temporary_32;
-	if (adamic_temporary_36 != NULL) adamic_release(adamic_temporary_36);
+	adamic_value *adamic_temporary_32 = (adamic_local_11_this->shape == &adamic_shape_0 || adamic_local_11_this->shape == &adamic_shape_1 ? (&adamic_local_11_this->slots[1]) : adamic_object_data_field(adamic_local_11_this, "#secret@1", &adamic_cache_34));
+	void *adamic_temporary_35 = adamic_temporary_32->reference;
+	adamic_temporary_32->reference = adamic_temporary_31;
+	if (adamic_temporary_35 != NULL) adamic_release(adamic_temporary_35);
 }
 
 static adamic_object * adamic_function_6_checked_view_members(adamic_object * adamic_local_15_argument) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_37 = ((adamic_string *)(&adamic_local_15_argument->slots[0])->reference);
-	(void)adamic_temporary_37;
-	adamic_object * adamic_temporary_38 = adamic_retain(adamic_local_15_argument);
-	return adamic_temporary_38;
+	adamic_string * adamic_temporary_36 = ((adamic_string *)(&adamic_local_15_argument->slots[0])->reference);
+	(void)adamic_temporary_36;
+	adamic_object * adamic_temporary_37 = adamic_retain(adamic_local_15_argument);
+	return adamic_temporary_37;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_7_ChildVault_new(adamic_string * adamic_local_24_value) {
 	ADAMIC_CHECK_STACK();
-	adamic_object * adamic_temporary_39 = adamic_object_new(&adamic_shape_1);
-	adamic_temporary_39->class = &adamic_class_2;
-	adamic_temporary_39->slots[0].reference = NULL;
-	adamic_temporary_39->slots[1].reference = NULL;
-	adamic_temporary_39->slots[2].reference = NULL;
-	adamic_temporary_39->slots[3].reference = NULL;
-	adamic_temporary_39->slots[4].reference = NULL;
-	adamic_object * adamic_local_23_this = adamic_temporary_39;
+	adamic_object * adamic_temporary_38 = adamic_object_new(&adamic_shape_1);
+	adamic_temporary_38->class = &adamic_class_2;
+	adamic_temporary_38->slots[0].reference = NULL;
+	adamic_temporary_38->slots[1].reference = NULL;
+	adamic_temporary_38->slots[2].reference = NULL;
+	adamic_temporary_38->slots[3].reference = NULL;
+	adamic_temporary_38->slots[4].reference = NULL;
+	adamic_object * adamic_local_23_this = adamic_temporary_38;
 	adamic_function_12_ChildVault_new_initialize(adamic_local_23_this, adamic_local_24_value);
-	adamic_object * adamic_temporary_40 = adamic_retain(adamic_local_23_this);
+	adamic_object * adamic_temporary_39 = adamic_retain(adamic_local_23_this);
 	adamic_release(adamic_local_23_this);
-	return adamic_temporary_40;
+	return adamic_temporary_39;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_7_ChildVault_new_in(adamic_region *region, adamic_string * adamic_local_24_value) {
 	ADAMIC_CHECK_STACK();
-	adamic_object *adamic_temporary_41 = adamic_object_new_in(region, &adamic_shape_1);
-	adamic_temporary_41->class = &adamic_class_2;
-	adamic_temporary_41->slots[0].reference = NULL;
-	adamic_temporary_41->slots[1].reference = NULL;
-	adamic_temporary_41->slots[2].reference = NULL;
-	adamic_temporary_41->slots[3].reference = NULL;
-	adamic_temporary_41->slots[4].reference = NULL;
-	adamic_object * adamic_local_23_this = adamic_retain(adamic_temporary_41);
+	adamic_object *adamic_temporary_40 = adamic_object_new_in(region, &adamic_shape_1);
+	adamic_temporary_40->class = &adamic_class_2;
+	adamic_temporary_40->slots[0].reference = NULL;
+	adamic_temporary_40->slots[1].reference = NULL;
+	adamic_temporary_40->slots[2].reference = NULL;
+	adamic_temporary_40->slots[3].reference = NULL;
+	adamic_temporary_40->slots[4].reference = NULL;
+	adamic_object * adamic_local_23_this = adamic_retain(adamic_temporary_40);
 	adamic_function_12_ChildVault_new_initialize(adamic_local_23_this, adamic_local_24_value);
-	adamic_object * adamic_temporary_42 = adamic_local_23_this;
+	adamic_object * adamic_temporary_41 = adamic_local_23_this;
 	adamic_release(adamic_local_23_this);
-	return adamic_temporary_42;
+	return adamic_temporary_41;
 	adamic_unreachable();
 }
 
 static void adamic_function_8_ChildVault_attach(adamic_object * adamic_local_16_this, adamic_object * adamic_local_17_parent) {
 	ADAMIC_CHECK_STACK();
-	adamic_weak * adamic_temporary_43 = adamic_weak_of(adamic_local_17_parent);
+	adamic_weak * adamic_temporary_42 = adamic_weak_of(adamic_local_17_parent);
 	if (adamic_local_16_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting '#parent@2')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_16_this, "#parent@2");
-	adamic_value *adamic_temporary_44 = (adamic_local_16_this->shape == &adamic_shape_1 ? (&adamic_local_16_this->slots[3]) : adamic_object_data_field(adamic_local_16_this, "#parent@2", &adamic_cache_46));
-	void *adamic_temporary_47 = adamic_temporary_44->reference;
-	adamic_temporary_44->reference = adamic_temporary_43;
-	if (adamic_temporary_47 != NULL) adamic_release(adamic_temporary_47);
+	adamic_value *adamic_temporary_43 = (adamic_local_16_this->shape == &adamic_shape_1 ? (&adamic_local_16_this->slots[3]) : adamic_object_data_field(adamic_local_16_this, "#parent@2", &adamic_cache_45));
+	void *adamic_temporary_46 = adamic_temporary_43->reference;
+	adamic_temporary_43->reference = adamic_temporary_42;
+	if (adamic_temporary_46 != NULL) adamic_release(adamic_temporary_46);
 	return;
 }
 
 static adamic_string * adamic_function_9_ChildVault_parentLabel(adamic_object * adamic_local_18_this) {
 	ADAMIC_CHECK_STACK();
-	adamic_weak * adamic_temporary_48 = adamic_retain(((adamic_weak *)(&adamic_local_18_this->slots[3])->reference));
-	adamic_object * adamic_temporary_49 = (adamic_object *)adamic_retain(adamic_weak_target(adamic_temporary_48));
-	adamic_weak * adamic_temporary_50 = adamic_weak_of(adamic_temporary_49);
-	adamic_weak * adamic_local_25_parent = adamic_temporary_50;
-	adamic_release(adamic_temporary_49);
+	adamic_weak * adamic_temporary_47 = adamic_retain(((adamic_weak *)(&adamic_local_18_this->slots[3])->reference));
+	adamic_object * adamic_temporary_48 = (adamic_object *)adamic_retain(adamic_weak_target(adamic_temporary_47));
+	adamic_weak * adamic_temporary_49 = adamic_weak_of(adamic_temporary_48);
+	adamic_weak * adamic_local_25_parent = adamic_temporary_49;
 	adamic_release(adamic_temporary_48);
-	adamic_object * adamic_temporary_51 = (adamic_object *)adamic_retain(adamic_weak_target(adamic_local_25_parent));
-	adamic_string * adamic_temporary_54;
-	if (adamic_temporary_51 == NULL) {
-		adamic_temporary_54 = adamic_retain(&adamic_string_4);
+	adamic_release(adamic_temporary_47);
+	adamic_object * adamic_temporary_50 = (adamic_object *)adamic_retain(adamic_weak_target(adamic_local_25_parent));
+	adamic_string * adamic_temporary_53;
+	if (adamic_temporary_50 == NULL) {
+		adamic_temporary_53 = adamic_retain(&adamic_string_4);
 	} else {
 		if (adamic_local_25_parent == NULL) {
 			static const char message[] = "TypeError: Cannot read properties of undefined (reading 'label')";
 			adamic_panic(message, sizeof message - 1);
 		}
-		adamic_object * adamic_temporary_52 = (adamic_object *)adamic_retain(adamic_weak_target_present(adamic_local_25_parent));
-		adamic_string * adamic_temporary_53 = adamic_retain(((adamic_string *)(&adamic_temporary_52->slots[0])->reference));
-		adamic_temporary_54 = adamic_retain(adamic_temporary_53);
-		adamic_release(adamic_temporary_53);
+		adamic_object * adamic_temporary_51 = (adamic_object *)adamic_retain(adamic_weak_target_present(adamic_local_25_parent));
+		adamic_string * adamic_temporary_52 = adamic_retain(((adamic_string *)(&adamic_temporary_51->slots[0])->reference));
+		adamic_temporary_53 = adamic_retain(adamic_temporary_52);
 		adamic_release(adamic_temporary_52);
-	}
-	adamic_string * adamic_temporary_55 = adamic_temporary_54;
-	adamic_release(adamic_temporary_51);
+		adamic_release(adamic_temporary_51);
+	}
+	adamic_string * adamic_temporary_54 = adamic_temporary_53;
+	adamic_release(adamic_temporary_50);
 	adamic_release(adamic_local_25_parent);
-	return adamic_temporary_55;
+	return adamic_temporary_54;
 	adamic_unreachable();
 }
 
 static adamic_string * adamic_function_10_ChildVault_read2(adamic_object * adamic_local_19_this) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_56 = ((adamic_string *)(&adamic_local_19_this->slots[4])->reference);
-	adamic_string * adamic_temporary_57 = adamic_retain(adamic_temporary_56);
-	return adamic_temporary_57;
+	adamic_string * adamic_temporary_55 = ((adamic_string *)(&adamic_local_19_this->slots[4])->reference);
+	adamic_string * adamic_temporary_56 = adamic_retain(adamic_temporary_55);
+	return adamic_temporary_56;
 	adamic_unreachable();
 }
 
 static adamic_string * adamic_function_11_ChildVault_readChild(adamic_object * adamic_local_20_this) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_58 = adamic_function_10_ChildVault_read2(adamic_local_20_this);
-	adamic_string * adamic_temporary_59 = adamic_temporary_58;
-	return adamic_temporary_59;
+	adamic_string * adamic_temporary_57 = adamic_function_10_ChildVault_read2(adamic_local_20_this);
+	adamic_string * adamic_temporary_58 = adamic_temporary_57;
+	return adamic_temporary_58;
 	adamic_unreachable();
 }
 
@@ -408,204 +406,204 @@
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_21_this, "child");
-	adamic_value *adamic_temporary_60 = adamic_object_data_field(adamic_local_21_this, "child", &adamic_cache_63);
-	void *adamic_temporary_64 = adamic_temporary_60->reference;
-	adamic_temporary_60->reference = adamic_retain(&adamic_string_3);
-	if (adamic_temporary_64 != NULL) adamic_release(adamic_temporary_64);
-	adamic_weak * adamic_temporary_65 = adamic_weak_of(NULL);
+	adamic_value *adamic_temporary_59 = adamic_object_data_field(adamic_local_21_this, "child", &adamic_cache_62);
+	void *adamic_temporary_63 = adamic_temporary_59->reference;
+	adamic_temporary_59->reference = adamic_retain(&adamic_string_3);
+	if (adamic_temporary_63 != NULL) adamic_release(adamic_temporary_63);
+	adamic_weak * adamic_temporary_64 = adamic_weak_of(NULL);
 	if (adamic_local_21_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting '#parent@2')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_21_this, "#parent@2");
-	adamic_value *adamic_temporary_66 = (adamic_local_21_this->shape == &adamic_shape_1 ? (&adamic_local_21_this->slots[3]) : adamic_object_data_field(adamic_local_21_this, "#parent@2", &adamic_cache_68));
-	void *adamic_temporary_69 = adamic_temporary_66->reference;
-	adamic_temporary_66->reference = adamic_temporary_65;
-	if (adamic_temporary_69 != NULL) adamic_release(adamic_temporary_69);
-	adamic_string * adamic_temporary_70 = adamic_string_from_number((0x1p+00));
-	adamic_string * adamic_temporary_71 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_3, adamic_temporary_70});
+	adamic_value *adamic_temporary_65 = (adamic_local_21_this->shape == &adamic_shape_1 ? (&adamic_local_21_this->slots[3]) : adamic_object_data_field(adamic_local_21_this, "#parent@2", &adamic_cache_67));
+	void *adamic_temporary_68 = adamic_temporary_65->reference;
+	adamic_temporary_65->reference = adamic_temporary_64;
+	if (adamic_temporary_68 != NULL) adamic_release(adamic_temporary_68);
+	adamic_string * adamic_temporary_69 = adamic_string_from_number((0x1p+00));
+	adamic_string * adamic_temporary_70 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_3, adamic_temporary_69});
 	if (adamic_local_21_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting '#secret@2')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_21_this, "#secret@2");
-	adamic_value *adamic_temporary_72 = (adamic_local_21_this->shape == &adamic_shape_1 ? (&adamic_local_21_this->slots[4]) : adamic_object_data_field(adamic_local_21_this, "#secret@2", &adamic_cache_74));
-	void *adamic_temporary_75 = adamic_temporary_72->reference;
-	adamic_temporary_72->reference = adamic_temporary_71;
-	if (adamic_temporary_75 != NULL) adamic_release(adamic_temporary_75);
-	adamic_release(adamic_temporary_70);
+	adamic_value *adamic_temporary_71 = (adamic_local_21_this->shape == &adamic_shape_1 ? (&adamic_local_21_this->slots[4]) : adamic_object_data_field(adamic_local_21_this, "#secret@2", &adamic_cache_73));
+	void *adamic_temporary_74 = adamic_temporary_71->reference;
+	adamic_temporary_71->reference = adamic_temporary_70;
+	if (adamic_temporary_74 != NULL) adamic_release(adamic_temporary_74);
+	adamic_release(adamic_temporary_69);
 }
 
 static adamic_object * adamic_function_13_checked_view_members(adamic_object * adamic_local_26_argument) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_77 = ((adamic_string *)adamic_object_data_field(adamic_local_26_argument, "child", &adamic_cache_76)->reference);
+	adamic_string * adamic_temporary_76 = ((adamic_string *)adamic_object_data_field(adamic_local_26_argument, "child", &adamic_cache_75)->reference);
+	(void)adamic_temporary_76;
+	adamic_string * adamic_temporary_77 = ((adamic_string *)(&adamic_local_26_argument->slots[0])->reference);
 	(void)adamic_temporary_77;
-	adamic_string * adamic_temporary_78 = ((adamic_string *)(&adamic_local_26_argument->slots[0])->reference);
-	(void)adamic_temporary_78;
-	adamic_object * adamic_temporary_79 = adamic_retain(adamic_local_26_argument);
-	return adamic_temporary_79;
+	adamic_object * adamic_temporary_78 = adamic_retain(adamic_local_26_argument);
+	return adamic_temporary_78;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_14_checked_view_members(adamic_object * adamic_local_27_argument) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_81 = ((adamic_string *)adamic_object_data_field(adamic_local_27_argument, "child", &adamic_cache_80)->reference);
+	adamic_string * adamic_temporary_80 = ((adamic_string *)adamic_object_data_field(adamic_local_27_argument, "child", &adamic_cache_79)->reference);
+	(void)adamic_temporary_80;
+	adamic_string * adamic_temporary_81 = ((adamic_string *)(&adamic_local_27_argument->slots[0])->reference);
 	(void)adamic_temporary_81;
-	adamic_string * adamic_temporary_82 = ((adamic_string *)(&adamic_local_27_argument->slots[0])->reference);
-	(void)adamic_temporary_82;
-	adamic_object * adamic_temporary_83 = adamic_retain(adamic_local_27_argument);
-	return adamic_temporary_83;
+	adamic_object * adamic_temporary_82 = adamic_retain(adamic_local_27_argument);
+	return adamic_temporary_82;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_15_checked_view_members(adamic_object * adamic_local_28_argument) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_84 = ((adamic_string *)(&adamic_local_28_argument->slots[0])->reference);
-	(void)adamic_temporary_84;
-	adamic_string * adamic_temporary_86 = ((adamic_string *)adamic_object_data_field(adamic_local_28_argument, "child", &adamic_cache_85)->reference);
-	(void)adamic_temporary_86;
-	adamic_object * adamic_temporary_87 = adamic_retain(adamic_local_28_argument);
-	return adamic_temporary_87;
+	adamic_string * adamic_temporary_83 = ((adamic_string *)(&adamic_local_28_argument->slots[0])->reference);
+	(void)adamic_temporary_83;
+	adamic_string * adamic_temporary_85 = ((adamic_string *)adamic_object_data_field(adamic_local_28_argument, "child", &adamic_cache_84)->reference);
+	(void)adamic_temporary_85;
+	adamic_object * adamic_temporary_86 = adamic_retain(adamic_local_28_argument);
+	return adamic_temporary_86;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_16_checked_view_members(adamic_object * adamic_local_29_argument) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_88 = ((adamic_string *)(&adamic_local_29_argument->slots[0])->reference);
-	(void)adamic_temporary_88;
-	adamic_object * adamic_temporary_89 = adamic_retain(adamic_local_29_argument);
-	return adamic_temporary_89;
+	adamic_string * adamic_temporary_87 = ((adamic_string *)(&adamic_local_29_argument->slots[0])->reference);
+	(void)adamic_temporary_87;
+	adamic_object * adamic_temporary_88 = adamic_retain(adamic_local_29_argument);
+	return adamic_temporary_88;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_17_checked_view_members(adamic_object * adamic_local_30_argument) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_90 = ((adamic_string *)(&adamic_local_30_argument->slots[0])->reference);
+	adamic_string * adamic_temporary_89 = ((adamic_string *)(&adamic_local_30_argument->slots[0])->reference);
+	(void)adamic_temporary_89;
+	adamic_string * adamic_temporary_90 = ((adamic_string *)(&adamic_local_30_argument->slots[1])->reference);
 	(void)adamic_temporary_90;
-	adamic_string * adamic_temporary_91 = ((adamic_string *)(&adamic_local_30_argument->slots[1])->reference);
-	(void)adamic_temporary_91;
-	adamic_object * adamic_temporary_92 = adamic_retain(adamic_local_30_argument);
-	return adamic_temporary_92;
+	adamic_object * adamic_temporary_91 = adamic_retain(adamic_local_30_argument);
+	return adamic_temporary_91;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_18_ReadonlyBox_new(void) {
 	ADAMIC_CHECK_STACK();
-	adamic_object * adamic_temporary_93 = adamic_object_new(&adamic_shape_2);
-	adamic_temporary_93->class = &adamic_class_3;
-	adamic_temporary_93->slots[0].reference = NULL;
-	adamic_temporary_93->slots[1].reference = NULL;
-	adamic_object * adamic_local_31_this = adamic_temporary_93;
+	adamic_object * adamic_temporary_92 = adamic_object_new(&adamic_shape_2);
+	adamic_temporary_92->class = &adamic_class_3;
+	adamic_temporary_92->slots[0].reference = NULL;
+	adamic_temporary_92->slots[1].reference = NULL;
+	adamic_object * adamic_local_31_this = adamic_temporary_92;
 	if (adamic_local_31_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'value')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_31_this, "value");
-	adamic_value *adamic_temporary_94 = adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_97);
-	void *adamic_temporary_98 = adamic_temporary_94->reference;
-	adamic_temporary_94->reference = NULL;
-	if (adamic_temporary_98 != NULL) adamic_release(adamic_temporary_98);
-	adamic_string * adamic_temporary_99 = adamic_string_from_number((0x1.8p+01));
-	adamic_string * adamic_temporary_100 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_12, adamic_temporary_99});
-	adamic_object * adamic_temporary_101 = adamic_object_new(&adamic_shape_3);
-	adamic_temporary_101->slots[0].reference = adamic_temporary_100;
+	adamic_value *adamic_temporary_93 = adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_96);
+	void *adamic_temporary_97 = adamic_temporary_93->reference;
+	adamic_temporary_93->reference = NULL;
+	if (adamic_temporary_97 != NULL) adamic_release(adamic_temporary_97);
+	adamic_string * adamic_temporary_98 = adamic_string_from_number((0x1.8p+01));
+	adamic_string * adamic_temporary_99 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_12, adamic_temporary_98});
+	adamic_object * adamic_temporary_100 = adamic_object_new(&adamic_shape_3);
+	adamic_temporary_100->slots[0].reference = adamic_temporary_99;
 	if (adamic_local_31_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'contents')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_31_this, "contents");
-	adamic_value *adamic_temporary_102 = (adamic_local_31_this->shape == &adamic_shape_2 ? (&adamic_local_31_this->slots[1]) : adamic_object_data_field(adamic_local_31_this, "contents", &adamic_cache_104));
-	void *adamic_temporary_105 = adamic_temporary_102->reference;
-	adamic_temporary_102->reference = adamic_temporary_101;
-	if (adamic_temporary_105 != NULL) adamic_release(adamic_temporary_105);
-	adamic_release(adamic_temporary_99);
-	adamic_string * adamic_temporary_106 = adamic_string_from_number((0x1p+02));
-	adamic_string * adamic_temporary_107 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_13, adamic_temporary_106});
+	adamic_value *adamic_temporary_101 = (adamic_local_31_this->shape == &adamic_shape_2 ? (&adamic_local_31_this->slots[1]) : adamic_object_data_field(adamic_local_31_this, "contents", &adamic_cache_103));
+	void *adamic_temporary_104 = adamic_temporary_101->reference;
+	adamic_temporary_101->reference = adamic_temporary_100;
+	if (adamic_temporary_104 != NULL) adamic_release(adamic_temporary_104);
+	adamic_release(adamic_temporary_98);
+	adamic_string * adamic_temporary_105 = adamic_string_from_number((0x1p+02));
+	adamic_string * adamic_temporary_106 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_13, adamic_temporary_105});
 	if (adamic_local_31_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'value')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_31_this, "value");
-	adamic_value *adamic_temporary_108 = (adamic_local_31_this->shape == &adamic_shape_2 ? &adamic_local_31_this->slots[0] : adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_111));
-	void *adamic_temporary_112 = adamic_temporary_108->reference;
-	adamic_temporary_108->reference = adamic_temporary_107;
-	if (adamic_temporary_112 != NULL) adamic_release(adamic_temporary_112);
-	adamic_release(adamic_temporary_106);
-	adamic_string * adamic_temporary_113 = adamic_string_from_number((0x1.4p+02));
-	adamic_string * adamic_temporary_114 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_14, adamic_temporary_113});
+	adamic_value *adamic_temporary_107 = (adamic_local_31_this->shape == &adamic_shape_2 ? &adamic_local_31_this->slots[0] : adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_110));
+	void *adamic_temporary_111 = adamic_temporary_107->reference;
+	adamic_temporary_107->reference = adamic_temporary_106;
+	if (adamic_temporary_111 != NULL) adamic_release(adamic_temporary_111);
+	adamic_release(adamic_temporary_105);
+	adamic_string * adamic_temporary_112 = adamic_string_from_number((0x1.4p+02));
+	adamic_string * adamic_temporary_113 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_14, adamic_temporary_112});
 	if (adamic_local_31_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'value')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_31_this, "value");
-	adamic_value *adamic_temporary_115 = (adamic_local_31_this->shape == &adamic_shape_2 ? &adamic_local_31_this->slots[0] : adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_118));
-	void *adamic_temporary_119 = adamic_temporary_115->reference;
-	adamic_temporary_115->reference = adamic_temporary_114;
-	if (adamic_temporary_119 != NULL) adamic_release(adamic_temporary_119);
-	adamic_release(adamic_temporary_113);
-	adamic_object * adamic_temporary_120 = adamic_retain(adamic_local_31_this);
+	adamic_value *adamic_temporary_114 = (adamic_local_31_this->shape == &adamic_shape_2 ? &adamic_local_31_this->slots[0] : adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_117));
+	void *adamic_temporary_118 = adamic_temporary_114->reference;
+	adamic_temporary_114->reference = adamic_temporary_113;
+	if (adamic_temporary_118 != NULL) adamic_release(adamic_temporary_118);
+	adamic_release(adamic_temporary_112);
+	adamic_object * adamic_temporary_119 = adamic_retain(adamic_local_31_this);
 	adamic_release(adamic_local_31_this);
-	return adamic_temporary_120;
+	return adamic_temporary_119;
 	adamic_unreachable();
 }
 
 static adamic_object * adamic_function_18_ReadonlyBox_new_in(adamic_region *region) {
 	ADAMIC_CHECK_STACK();
-	adamic_object *adamic_temporary_121 = adamic_object_new_in(region, &adamic_shape_2);
-	adamic_temporary_121->class = &adamic_class_3;
-	adamic_temporary_121->slots[0].reference = NULL;
-	adamic_temporary_121->slots[1].reference = NULL;
-	adamic_object * adamic_local_31_this = adamic_retain(adamic_temporary_121);
+	adamic_object *adamic_temporary_120 = adamic_object_new_in(region, &adamic_shape_2);
+	adamic_temporary_120->class = &adamic_class_3;
+	adamic_temporary_120->slots[0].reference = NULL;
+	adamic_temporary_120->slots[1].reference = NULL;
+	adamic_object * adamic_local_31_this = adamic_retain(adamic_temporary_120);
 	if (adamic_local_31_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'value')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_31_this, "value");
-	adamic_value *adamic_temporary_122 = adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_125);
-	void *adamic_temporary_126 = adamic_temporary_122->reference;
-	adamic_temporary_122->reference = NULL;
-	if (adamic_temporary_126 != NULL) adamic_release(adamic_temporary_126);
-	adamic_string * adamic_temporary_127 = adamic_string_from_number((0x1.8p+01));
-	adamic_string * adamic_temporary_128 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_12, adamic_temporary_127});
-	adamic_object * adamic_temporary_129 = adamic_object_new(&adamic_shape_3);
-	adamic_temporary_129->slots[0].reference = adamic_temporary_128;
+	adamic_value *adamic_temporary_121 = adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_124);
+	void *adamic_temporary_125 = adamic_temporary_121->reference;
+	adamic_temporary_121->reference = NULL;
+	if (adamic_temporary_125 != NULL) adamic_release(adamic_temporary_125);
+	adamic_string * adamic_temporary_126 = adamic_string_from_number((0x1.8p+01));
+	adamic_string * adamic_temporary_127 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_12, adamic_temporary_126});
+	adamic_object * adamic_temporary_128 = adamic_object_new(&adamic_shape_3);
+	adamic_temporary_128->slots[0].reference = adamic_temporary_127;
 	if (adamic_local_31_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'contents')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_31_this, "contents");
-	adamic_value *adamic_temporary_130 = (adamic_local_31_this->shape == &adamic_shape_2 ? (&adamic_local_31_this->slots[1]) : adamic_object_data_field(adamic_local_31_this, "contents", &adamic_cache_132));
-	void *adamic_temporary_133 = adamic_temporary_130->reference;
-	adamic_temporary_130->reference = adamic_temporary_129;
-	if (adamic_temporary_133 != NULL) adamic_release(adamic_temporary_133);
-	adamic_release(adamic_temporary_127);
-	adamic_string * adamic_temporary_134 = adamic_string_from_number((0x1p+02));
-	adamic_string * adamic_temporary_135 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_13, adamic_temporary_134});
+	adamic_value *adamic_temporary_129 = (adamic_local_31_this->shape == &adamic_shape_2 ? (&adamic_local_31_this->slots[1]) : adamic_object_data_field(adamic_local_31_this, "contents", &adamic_cache_131));
+	void *adamic_temporary_132 = adamic_temporary_129->reference;
+	adamic_temporary_129->reference = adamic_temporary_128;
+	if (adamic_temporary_132 != NULL) adamic_release(adamic_temporary_132);
+	adamic_release(adamic_temporary_126);
+	adamic_string * adamic_temporary_133 = adamic_string_from_number((0x1p+02));
+	adamic_string * adamic_temporary_134 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_13, adamic_temporary_133});
 	if (adamic_local_31_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'value')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_31_this, "value");
-	adamic_value *adamic_temporary_136 = (adamic_local_31_this->shape == &adamic_shape_2 ? &adamic_local_31_this->slots[0] : adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_139));
-	void *adamic_temporary_140 = adamic_temporary_136->reference;
-	adamic_temporary_136->reference = adamic_temporary_135;
-	if (adamic_temporary_140 != NULL) adamic_release(adamic_temporary_140);
-	adamic_release(adamic_temporary_134);
-	adamic_string * adamic_temporary_141 = adamic_string_from_number((0x1.4p+02));
-	adamic_string * adamic_temporary_142 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_14, adamic_temporary_141});
+	adamic_value *adamic_temporary_135 = (adamic_local_31_this->shape == &adamic_shape_2 ? &adamic_local_31_this->slots[0] : adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_138));
+	void *adamic_temporary_139 = adamic_temporary_135->reference;
+	adamic_temporary_135->reference = adamic_temporary_134;
+	if (adamic_temporary_139 != NULL) adamic_release(adamic_temporary_139);
+	adamic_release(adamic_temporary_133);
+	adamic_string * adamic_temporary_140 = adamic_string_from_number((0x1.4p+02));
+	adamic_string * adamic_temporary_141 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_14, adamic_temporary_140});
 	if (adamic_local_31_this == NULL) {
 		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'value')";
 		adamic_panic(message, sizeof message - 1);
 	}
 	adamic_object_check_data_write(adamic_local_31_this, "value");
-	adamic_value *adamic_temporary_143 = (adamic_local_31_this->shape == &adamic_shape_2 ? &adamic_local_31_this->slots[0] : adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_146));
-	void *adamic_temporary_147 = adamic_temporary_143->reference;
-	adamic_temporary_143->reference = adamic_temporary_142;
-	if (adamic_temporary_147 != NULL) adamic_release(adamic_temporary_147);
-	adamic_release(adamic_temporary_141);
-	adamic_object * adamic_temporary_148 = adamic_local_31_this;
+	adamic_value *adamic_temporary_142 = (adamic_local_31_this->shape == &adamic_shape_2 ? &adamic_local_31_this->slots[0] : adamic_object_data_field(adamic_local_31_this, "value", &adamic_cache_145));
+	void *adamic_temporary_146 = adamic_temporary_142->reference;
+	adamic_temporary_142->reference = adamic_temporary_141;
+	if (adamic_temporary_146 != NULL) adamic_release(adamic_temporary_146);
+	adamic_release(adamic_temporary_140);
+	adamic_object * adamic_temporary_147 = adamic_local_31_this;
 	adamic_release(adamic_local_31_this);
-	return adamic_temporary_148;
+	return adamic_temporary_147;
 	adamic_unreachable();
 }
 
@@ -615,132 +613,126 @@
 
 int main(int argc, char **argv) {
 	adamic_start(argc, argv);
-	adamic_object * adamic_temporary_149 = adamic_function_7_ChildVault_new(&adamic_string_5);
+	adamic_object * adamic_temporary_148 = adamic_function_7_ChildVault_new(&adamic_string_5);
+	adamic_object * adamic_temporary_149 = adamic_global_0_child;
+	adamic_global_0_child = adamic_temporary_148;
+	adamic_release(adamic_temporary_149);
+	adamic_ready_0 = true;
 	adamic_object * adamic_temporary_150 = adamic_global_0_child;
-	adamic_global_0_child = adamic_temporary_149;
-	adamic_release(adamic_temporary_150);
-	adamic_ready_0 = true;
-	adamic_object * adamic_temporary_151 = adamic_global_0_child;
-	adamic_string * adamic_temporary_152 = adamic_function_3_Vault_read(adamic_temporary_151);
-	adamic_object * adamic_temporary_153 = adamic_global_0_child;
-	adamic_string * adamic_temporary_154 = adamic_function_11_ChildVault_readChild(adamic_temporary_153);
-	adamic_string * adamic_temporary_155 = adamic_string_concat(3, (adamic_string *const[]){adamic_temporary_152, &adamic_string_6, adamic_temporary_154});
-	adamic_write_line(adamic_stdout, adamic_temporary_155);
-	adamic_release(adamic_temporary_155);
+	adamic_string * adamic_temporary_151 = adamic_function_3_Vault_read(adamic_temporary_150);
+	adamic_object * adamic_temporary_152 = adamic_global_0_child;
+	adamic_string * adamic_temporary_153 = adamic_function_11_ChildVault_readChild(adamic_temporary_152);
+	adamic_string * adamic_temporary_154 = adamic_string_concat(3, (adamic_string *const[]){adamic_temporary_151, &adamic_string_6, adamic_temporary_153});
+	adamic_write_line(adamic_stdout, adamic_temporary_154);
 	adamic_release(adamic_temporary_154);
-	adamic_release(adamic_temporary_152);
+	adamic_release(adamic_temporary_153);
+	adamic_release(adamic_temporary_151);
+	adamic_object * adamic_temporary_155 = adamic_global_0_child;
+	adamic_function_4_Vault_replace(adamic_temporary_155, &adamic_string_7);
 	adamic_object * adamic_temporary_156 = adamic_global_0_child;
-	adamic_function_4_Vault_replace(adamic_temporary_156, &adamic_string_7);
-	adamic_object * adamic_temporary_157 = adamic_global_0_child;
-	adamic_string * adamic_temporary_158 = adamic_function_3_Vault_read(adamic_temporary_157);
-	adamic_object * adamic_temporary_159 = adamic_global_0_child;
-	adamic_string * adamic_temporary_160 = adamic_function_11_ChildVault_readChild(adamic_temporary_159);
-	adamic_string * adamic_temporary_161 = adamic_string_concat(3, (adamic_string *const[]){adamic_temporary_158, &adamic_string_6, adamic_temporary_160});
-	adamic_write_line(adamic_stdout, adamic_temporary_161);
+	adamic_string * adamic_temporary_157 = adamic_function_3_Vault_read(adamic_temporary_156);
+	adamic_object * adamic_temporary_158 = adamic_global_0_child;
+	adamic_string * adamic_temporary_159 = adamic_function_11_ChildVault_readChild(adamic_temporary_158);
+	adamic_string * adamic_temporary_160 = adamic_string_concat(3, (adamic_string *const[]){adamic_temporary_157, &adamic_string_6, adamic_temporary_159});
+	adamic_write_line(adamic_stdout, adamic_temporary_160);
+	adamic_release(adamic_temporary_160);
+	adamic_release(adamic_temporary_159);
+	adamic_release(adamic_temporary_157);
+	adamic_object * adamic_temporary_161 = adamic_retain(adamic_global_0_child);
+	adamic_array * adamic_temporary_162 = adamic_class_object_keys(adamic_temporary_161);
+	adamic_string * adamic_temporary_163 = adamic_array_join(adamic_temporary_162, &adamic_string_8, adamic_join_strings);
+	adamic_write_line(adamic_stdout, adamic_temporary_163);
+	adamic_release(adamic_temporary_163);
+	adamic_release(adamic_temporary_162);
 	adamic_release(adamic_temporary_161);
-	adamic_release(adamic_temporary_160);
-	adamic_release(adamic_temporary_158);
-	adamic_object * adamic_temporary_162 = adamic_global_0_child;
-	adamic_object * adamic_temporary_163 = adamic_function_13_checked_view_members(adamic_temporary_162);
-	adamic_array * adamic_temporary_164 = adamic_class_object_keys(adamic_temporary_163);
-	adamic_string * adamic_temporary_165 = adamic_array_join(adamic_temporary_164, &adamic_string_8, adamic_join_strings);
-	adamic_write_line(adamic_stdout, adamic_temporary_165);
+	adamic_object * adamic_temporary_164 = adamic_retain(adamic_global_0_child);
+	adamic_object * adamic_temporary_165 = adamic_retain(adamic_temporary_164);
+	adamic_object * adamic_temporary_166 = adamic_object_copy(adamic_temporary_165);
+	adamic_object * adamic_temporary_167 = adamic_global_1_copied;
+	adamic_global_1_copied = adamic_temporary_166;
+	adamic_release(adamic_temporary_167);
+	adamic_ready_1 = true;
 	adamic_release(adamic_temporary_165);
 	adamic_release(adamic_temporary_164);
-	adamic_release(adamic_temporary_163);
-	adamic_object * adamic_temporary_166 = adamic_global_0_child;
-	adamic_object * adamic_temporary_167 = adamic_function_14_checked_view_members(adamic_temporary_166);
-	adamic_object * adamic_temporary_168 = adamic_retain(adamic_temporary_167);
-	adamic_object * adamic_temporary_169 = adamic_object_copy(adamic_temporary_168);
-	adamic_object * adamic_temporary_170 = adamic_global_1_copied;
-	adamic_global_1_copied = adamic_temporary_169;
+	adamic_object * adamic_temporary_168 = adamic_retain(adamic_global_1_copied);
+	adamic_array * adamic_temporary_169 = adamic_class_object_keys(adamic_temporary_168);
+	adamic_string * adamic_temporary_170 = adamic_array_join(adamic_temporary_169, &adamic_string_8, adamic_join_strings);
+	adamic_object * adamic_temporary_171 = adamic_global_1_copied;
+	adamic_string * adamic_temporary_172 = adamic_retain(((adamic_string *)(&adamic_temporary_171->slots[0])->reference));
+	adamic_object * adamic_temporary_173 = adamic_global_1_copied;
+	adamic_string * adamic_temporary_175 = adamic_retain(((adamic_string *)(adamic_temporary_173->shape == &adamic_shape_1 ? &adamic_temporary_173->slots[2] : adamic_object_data_field(adamic_temporary_173, "child", &adamic_cache_174))->reference));
+	adamic_string * adamic_temporary_176 = adamic_string_concat(5, (adamic_string *const[]){adamic_temporary_170, &adamic_string_6, adamic_temporary_172, &adamic_string_6, adamic_temporary_175});
+	adamic_write_line(adamic_stdout, adamic_temporary_176);
+	adamic_release(adamic_temporary_176);
+	adamic_release(adamic_temporary_175);
+	adamic_release(adamic_temporary_172);
 	adamic_release(adamic_temporary_170);
-	adamic_ready_1 = true;
+	adamic_release(adamic_temporary_169);
 	adamic_release(adamic_temporary_168);
-	adamic_release(adamic_temporary_167);
-	adamic_object * adamic_temporary_171 = adamic_global_1_copied;
-	adamic_object * adamic_temporary_172 = adamic_function_15_checked_view_members(adamic_temporary_171);
-	adamic_array * adamic_temporary_173 = adamic_class_object_keys(adamic_temporary_172);
-	adamic_string * adamic_temporary_174 = adamic_array_join(adamic_temporary_173, &adamic_string_8, adamic_join_strings);
-	adamic_object * adamic_temporary_175 = adamic_global_1_copied;
-	adamic_string * adamic_temporary_176 = adamic_retain(((adamic_string *)(&adamic_temporary_175->slots[0])->reference));
-	adamic_object * adamic_temporary_177 = adamic_global_1_copied;
-	adamic_string * adamic_temporary_179 = adamic_retain(((adamic_string *)(adamic_temporary_177->shape == &adamic_shape_1 ? &adamic_temporary_177->slots[2] : adamic_object_data_field(adamic_temporary_177, "child", &adamic_cache_178))->reference));
-	adamic_string * adamic_temporary_180 = adamic_string_concat(5, (adamic_string *const[]){adamic_temporary_174, &adamic_string_6, adamic_temporary_176, &adamic_string_6, adamic_temporary_179});
-	adamic_write_line(adamic_stdout, adamic_temporary_180);
+	adamic_object * adamic_temporary_177 = adamic_function_0_copy(&adamic_string_9);
+	adamic_array * adamic_temporary_178 = adamic_class_object_keys(adamic_temporary_177);
+	adamic_string * adamic_temporary_179 = adamic_array_join(adamic_temporary_178, &adamic_string_8, adamic_join_strings);
+	adamic_write_line(adamic_stdout, adamic_temporary_179);
+	adamic_release(adamic_temporary_179);
+	adamic_release(adamic_temporary_178);
+	adamic_release(adamic_temporary_177);
+	adamic_string * adamic_temporary_180 = adamic_string_from_number((0x1p+01));
+	adamic_string * adamic_temporary_181 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_10, adamic_temporary_180});
+	adamic_object * adamic_temporary_182 = adamic_object_new(&adamic_shape_4);
+	adamic_temporary_182->slots[0].reference = adamic_temporary_181;
+	adamic_temporary_182->slots[1].reference = adamic_retain(&adamic_string_11);
+	adamic_object * adamic_temporary_183 = adamic_global_2_ordinary;
+	adamic_global_2_ordinary = adamic_temporary_182;
+	adamic_release(adamic_temporary_183);
+	adamic_ready_2 = true;
 	adamic_release(adamic_temporary_180);
-	adamic_release(adamic_temporary_179);
-	adamic_release(adamic_temporary_176);
-	adamic_release(adamic_temporary_174);
-	adamic_release(adamic_temporary_173);
-	adamic_release(adamic_temporary_172);
-	adamic_object * adamic_temporary_181 = adamic_function_0_copy(&adamic_string_9);
-	adamic_object * adamic_temporary_182 = adamic_function_16_checked_view_members(adamic_temporary_181);
-	adamic_array * adamic_temporary_183 = adamic_class_object_keys(adamic_temporary_182);
-	adamic_string * adamic_temporary_184 = adamic_array_join(adamic_temporary_183, &adamic_string_8, adamic_join_strings);
-	adamic_write_line(adamic_stdout, adamic_temporary_184);
+	adamic_object * adamic_temporary_184 = adamic_retain(adamic_global_2_ordinary);
+	adamic_array * adamic_temporary_185 = adamic_class_object_keys(adamic_temporary_184);
+	adamic_string * adamic_temporary_186 = adamic_array_join(adamic_temporary_185, &adamic_string_8, adamic_join_strings);
+	adamic_write_line(adamic_stdout, adamic_temporary_186);
+	adamic_release(adamic_temporary_186);
+	adamic_release(adamic_temporary_185);
 	adamic_release(adamic_temporary_184);
-	adamic_release(adamic_temporary_183);
-	adamic_release(adamic_temporary_182);
-	adamic_release(adamic_temporary_181);
-	adamic_string * adamic_temporary_185 = adamic_string_from_number((0x1p+01));
-	adamic_string * adamic_temporary_186 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_10, adamic_temporary_185});
-	adamic_object * adamic_temporary_187 = adamic_object_new(&adamic_shape_4);
-	adamic_temporary_187->slots[0].reference = adamic_temporary_186;
-	adamic_temporary_187->slots[1].reference = adamic_retain(&adamic_string_11);
-	adamic_object * adamic_temporary_188 = adamic_global_2_ordinary;
-	adamic_global_2_ordinary = adamic_temporary_187;
+	adamic_object * adamic_temporary_187 = adamic_function_18_ReadonlyBox_new();
+	adamic_object * adamic_temporary_188 = adamic_global_3_readonlyBox;
+	adamic_global_3_readonlyBox = adamic_temporary_187;
 	adamic_release(adamic_temporary_188);
-	adamic_ready_2 = true;
-	adamic_release(adamic_temporary_185);
-	adamic_object * adamic_temporary_189 = adamic_global_2_ordinary;
-	adamic_object * adamic_temporary_190 = adamic_function_17_checked_view_members(adamic_temporary_189);
-	adamic_array * adamic_temporary_191 = adamic_class_object_keys(adamic_temporary_190);
-	adamic_string * adamic_temporary_192 = adamic_array_join(adamic_temporary_191, &adamic_string_8, adamic_join_strings);
-	adamic_write_line(adamic_stdout, adamic_temporary_192);
-	adamic_release(adamic_temporary_192);
+	adamic_ready_3 = true;
+	adamic_object * adamic_temporary_189 = adamic_global_3_readonlyBox;
+	adamic_object * adamic_temporary_190 = adamic_retain(((adamic_object *)(&adamic_temporary_189->slots[1])->reference));
+	adamic_string * adamic_temporary_191 = adamic_string_from_number((0x1.8p+02));
+	adamic_string * adamic_temporary_192 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_2, adamic_temporary_191});
+	if (adamic_temporary_190 == NULL) {
+		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'text')";
+		adamic_panic(message, sizeof message - 1);
+	}
+	adamic_object_check_data_write(adamic_temporary_190, "text");
+	adamic_value *adamic_temporary_193 = adamic_object_data_field(adamic_temporary_190, "text", &adamic_cache_196);
+	void *adamic_temporary_197 = adamic_temporary_193->reference;
+	adamic_temporary_193->reference = adamic_temporary_192;
+	if (adamic_temporary_197 != NULL) adamic_release(adamic_temporary_197);
 	adamic_release(adamic_temporary_191);
 	adamic_release(adamic_temporary_190);
-	adamic_object * adamic_temporary_193 = adamic_function_18_ReadonlyBox_new();
-	adamic_object * adamic_temporary_194 = adamic_global_3_readonlyBox;
-	adamic_global_3_readonlyBox = adamic_temporary_193;
-	adamic_release(adamic_temporary_194);
-	adamic_ready_3 = true;
-	adamic_object * adamic_temporary_195 = adamic_global_3_readonlyBox;
-	adamic_object * adamic_temporary_196 = adamic_retain(((adamic_object *)(&adamic_temporary_195->slots[1])->reference));
-	adamic_string * adamic_temporary_197 = adamic_string_from_number((0x1.8p+02));
-	adamic_string * adamic_temporary_198 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_2, adamic_temporary_197});
-	if (adamic_temporary_196 == NULL) {
-		static const char message[] = "TypeError: Cannot set properties of undefined (setting 'text')";
-		adamic_panic(message, sizeof message - 1);
-	}
-	adamic_object_check_data_write(adamic_temporary_196, "text");
-	adamic_value *adamic_temporary_199 = adamic_object_data_field(adamic_temporary_196, "text", &adamic_cache_202);
-	void *adamic_temporary_203 = adamic_temporary_199->reference;
-	adamic_temporary_199->reference = adamic_temporary_198;
-	if (adamic_temporary_203 != NULL) adamic_release(adamic_temporary_203);
-	adamic_release(adamic_temporary_197);
-	adamic_release(adamic_temporary_196);
-	adamic_object * adamic_temporary_204 = adamic_global_3_readonlyBox;
-	adamic_string * adamic_temporary_206 = ((adamic_string *)(adamic_temporary_204->shape == &adamic_shape_2 ? &adamic_temporary_204->slots[0] : adamic_object_data_field(adamic_temporary_204, "value", &adamic_cache_205))->reference);
-	adamic_object * adamic_temporary_207 = adamic_global_3_readonlyBox;
-	adamic_object * adamic_temporary_208 = ((adamic_object *)(&adamic_temporary_207->slots[1])->reference);
-	adamic_string * adamic_temporary_210 = ((adamic_string *)adamic_object_data_field(adamic_temporary_208, "text", &adamic_cache_209)->reference);
-	adamic_string * adamic_temporary_211 = adamic_string_concat(3, (adamic_string *const[]){adamic_temporary_206, &adamic_string_6, adamic_temporary_210});
+	adamic_object * adamic_temporary_198 = adamic_global_3_readonlyBox;
+	adamic_string * adamic_temporary_200 = ((adamic_string *)(adamic_temporary_198->shape == &adamic_shape_2 ? &adamic_temporary_198->slots[0] : adamic_object_data_field(adamic_temporary_198, "value", &adamic_cache_199))->reference);
+	adamic_object * adamic_temporary_201 = adamic_global_3_readonlyBox;
+	adamic_object * adamic_temporary_202 = ((adamic_object *)(&adamic_temporary_201->slots[1])->reference);
+	adamic_string * adamic_temporary_204 = ((adamic_string *)adamic_object_data_field(adamic_temporary_202, "text", &adamic_cache_203)->reference);
+	adamic_string * adamic_temporary_205 = adamic_string_concat(3, (adamic_string *const[]){adamic_temporary_200, &adamic_string_6, adamic_temporary_204});
+	adamic_write_line(adamic_stdout, adamic_temporary_205);
+	adamic_release(adamic_temporary_205);
+	adamic_object * adamic_temporary_206 = adamic_function_1_Vault_new(&adamic_string_15);
+	adamic_object * adamic_temporary_207 = adamic_global_4_parent;
+	adamic_global_4_parent = adamic_temporary_206;
+	adamic_release(adamic_temporary_207);
+	adamic_ready_4 = true;
+	adamic_object * adamic_temporary_208 = adamic_global_0_child;
+	adamic_object * adamic_temporary_209 = adamic_global_4_parent;
+	adamic_function_8_ChildVault_attach(adamic_temporary_208, adamic_temporary_209);
+	adamic_object * adamic_temporary_210 = adamic_global_0_child;
+	adamic_string * adamic_temporary_211 = adamic_function_9_ChildVault_parentLabel(adamic_temporary_210);
 	adamic_write_line(adamic_stdout, adamic_temporary_211);
 	adamic_release(adamic_temporary_211);
-	adamic_object * adamic_temporary_212 = adamic_function_1_Vault_new(&adamic_string_15);
-	adamic_object * adamic_temporary_213 = adamic_global_4_parent;
-	adamic_global_4_parent = adamic_temporary_212;
-	adamic_release(adamic_temporary_213);
-	adamic_ready_4 = true;
-	adamic_object * adamic_temporary_214 = adamic_global_0_child;
-	adamic_object * adamic_temporary_215 = adamic_global_4_parent;
-	adamic_function_8_ChildVault_attach(adamic_temporary_214, adamic_temporary_215);
-	adamic_object * adamic_temporary_216 = adamic_global_0_child;
-	adamic_string * adamic_temporary_217 = adamic_function_9_ChildVault_parentLabel(adamic_temporary_216);
-	adamic_write_line(adamic_stdout, adamic_temporary_217);
-	adamic_release(adamic_temporary_217);
 	adamic_release(adamic_global_4_parent);
 	adamic_global_4_parent = NULL;
 	adamic_release(adamic_global_3_readonlyBox);
