--- 51aa3a96
+++ 6a056842
@@ -14,11 +14,11 @@
 static void adamic_function_1_run(void);
 static adamic_object * adamic_function_2_checked_view_members(adamic_object * adamic_local_3_argument);
 
-static adamic_slot_cache adamic_cache_6;
+static adamic_slot_cache adamic_cache_5;
 static const char *const adamic_shape_0_names[] = {"label"};
 static const bool adamic_shape_0_references[] = {true};
 static const adamic_shape adamic_shape_0 = {1, adamic_shape_0_names, adamic_shape_0_references, NULL};
-static int adamic_compare_19(adamic_value left, adamic_value right, void *context) {
+static int adamic_compare_18(adamic_value left, adamic_value right, void *context) {
 	(void)context;
 	double result = adamic_function_0_compare(left.reference, right.reference);
 	return result < 0 ? -1 : result > 0 ? 1 : 0;
@@ -26,73 +26,71 @@
 
 static double adamic_function_0_compare(adamic_object * adamic_local_0_left, adamic_object * adamic_local_1_right) {
 	ADAMIC_CHECK_STACK();
-	adamic_object * adamic_temporary_1 = adamic_function_2_checked_view_members(adamic_local_0_left);
-	adamic_object * adamic_temporary_2 = adamic_object_copy(adamic_temporary_1);
-	adamic_string * adamic_temporary_3 = ((adamic_string *)(&adamic_local_0_left->slots[0])->reference);
-	adamic_string * adamic_temporary_4 = adamic_string_concat(2, (adamic_string *const[]){adamic_temporary_3, &adamic_string_0});
-	adamic_value *adamic_temporary_5 = adamic_object_field(adamic_temporary_2, "label", &adamic_cache_6);
-	adamic_release(adamic_temporary_5->reference);
-	adamic_temporary_5->reference = adamic_temporary_4;
-	adamic_object * adamic_local_2_copy = adamic_temporary_2;
-	adamic_release(adamic_temporary_1);
-	adamic_string * adamic_temporary_7 = ((adamic_string *)(&adamic_local_2_copy->slots[0])->reference);
-	double adamic_temporary_8 = adamic_string_length(adamic_temporary_7);
-	adamic_string * adamic_temporary_9 = ((adamic_string *)(&adamic_local_1_right->slots[0])->reference);
-	double adamic_temporary_10 = adamic_string_length(adamic_temporary_9);
-	double adamic_temporary_11 = (adamic_temporary_8 - adamic_temporary_10);
+	adamic_object * adamic_temporary_1 = adamic_object_copy(adamic_local_0_left);
+	adamic_string * adamic_temporary_2 = ((adamic_string *)(&adamic_local_0_left->slots[0])->reference);
+	adamic_string * adamic_temporary_3 = adamic_string_concat(2, (adamic_string *const[]){adamic_temporary_2, &adamic_string_0});
+	adamic_value *adamic_temporary_4 = adamic_object_field(adamic_temporary_1, "label", &adamic_cache_5);
+	adamic_release(adamic_temporary_4->reference);
+	adamic_temporary_4->reference = adamic_temporary_3;
+	adamic_object * adamic_local_2_copy = adamic_temporary_1;
+	adamic_string * adamic_temporary_6 = ((adamic_string *)(&adamic_local_2_copy->slots[0])->reference);
+	double adamic_temporary_7 = adamic_string_length(adamic_temporary_6);
+	adamic_string * adamic_temporary_8 = ((adamic_string *)(&adamic_local_1_right->slots[0])->reference);
+	double adamic_temporary_9 = adamic_string_length(adamic_temporary_8);
+	double adamic_temporary_10 = (adamic_temporary_7 - adamic_temporary_9);
 	adamic_release(adamic_local_2_copy);
-	return adamic_temporary_11;
+	return adamic_temporary_10;
 	adamic_unreachable();
 }
 
 static void adamic_function_1_run(void) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_12 = adamic_string_from_number((0x1p+00));
-	adamic_string * adamic_temporary_13 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_1, adamic_temporary_12});
-	adamic_object * adamic_temporary_14 = adamic_object_new(&adamic_shape_0);
-	adamic_temporary_14->slots[0].reference = adamic_temporary_13;
-	adamic_string * adamic_temporary_15 = adamic_string_from_number((0x1.6p+04));
-	adamic_string * adamic_temporary_16 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_2, adamic_temporary_15});
-	adamic_object * adamic_temporary_17 = adamic_object_new(&adamic_shape_0);
-	adamic_temporary_17->slots[0].reference = adamic_temporary_16;
-	adamic_array * adamic_temporary_18 = adamic_array_new(2, true);
-	adamic_array_push(adamic_temporary_18, (adamic_value){.reference = adamic_retain(adamic_temporary_14)});
-	adamic_array_push(adamic_temporary_18, (adamic_value){.reference = adamic_retain(adamic_temporary_17)});
-	adamic_array * adamic_local_4_items = adamic_temporary_18;
-	adamic_release(adamic_temporary_17);
-	adamic_release(adamic_temporary_15);
+	adamic_string * adamic_temporary_11 = adamic_string_from_number((0x1p+00));
+	adamic_string * adamic_temporary_12 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_1, adamic_temporary_11});
+	adamic_object * adamic_temporary_13 = adamic_object_new(&adamic_shape_0);
+	adamic_temporary_13->slots[0].reference = adamic_temporary_12;
+	adamic_string * adamic_temporary_14 = adamic_string_from_number((0x1.6p+04));
+	adamic_string * adamic_temporary_15 = adamic_string_concat(2, (adamic_string *const[]){&adamic_string_2, adamic_temporary_14});
+	adamic_object * adamic_temporary_16 = adamic_object_new(&adamic_shape_0);
+	adamic_temporary_16->slots[0].reference = adamic_temporary_15;
+	adamic_array * adamic_temporary_17 = adamic_array_new(2, true);
+	adamic_array_push(adamic_temporary_17, (adamic_value){.reference = adamic_retain(adamic_temporary_13)});
+	adamic_array_push(adamic_temporary_17, (adamic_value){.reference = adamic_retain(adamic_temporary_16)});
+	adamic_array * adamic_local_4_items = adamic_temporary_17;
+	adamic_release(adamic_temporary_16);
 	adamic_release(adamic_temporary_14);
-	adamic_release(adamic_temporary_12);
-	adamic_array_sort(adamic_local_4_items, adamic_compare_19, NULL);
+	adamic_release(adamic_temporary_13);
+	adamic_release(adamic_temporary_11);
+	adamic_array_sort(adamic_local_4_items, adamic_compare_18, NULL);
 	(void)adamic_local_4_items;
-	adamic_value *adamic_temporary_20 = adamic_array_at(adamic_local_4_items, (0x0p+00));
-	adamic_object * adamic_temporary_21 = adamic_temporary_20 == NULL ? NULL : (adamic_object *)adamic_retain(adamic_temporary_20->reference);
-	adamic_string * adamic_temporary_22 = adamic_retain((adamic_temporary_21 == NULL ? NULL : ((adamic_string *)(&adamic_temporary_21->slots[0])->reference)));
-	adamic_string * adamic_temporary_23;
-	if (adamic_temporary_22 != NULL) {
-		adamic_temporary_23 = adamic_temporary_22;
+	adamic_value *adamic_temporary_19 = adamic_array_at(adamic_local_4_items, (0x0p+00));
+	adamic_object * adamic_temporary_20 = adamic_temporary_19 == NULL ? NULL : (adamic_object *)adamic_retain(adamic_temporary_19->reference);
+	adamic_string * adamic_temporary_21 = adamic_retain((adamic_temporary_20 == NULL ? NULL : ((adamic_string *)(&adamic_temporary_20->slots[0])->reference)));
+	adamic_string * adamic_temporary_22;
+	if (adamic_temporary_21 != NULL) {
+		adamic_temporary_22 = adamic_temporary_21;
 	} else {
-		adamic_temporary_23 = adamic_retain(&adamic_string_3);
+		adamic_temporary_22 = adamic_retain(&adamic_string_3);
 	}
-	adamic_string * adamic_temporary_24 = adamic_string_concat(1, (adamic_string *const[]){adamic_temporary_23});
-	adamic_write_line(adamic_stdout, adamic_temporary_24);
-	adamic_release(adamic_temporary_24);
+	adamic_string * adamic_temporary_23 = adamic_string_concat(1, (adamic_string *const[]){adamic_temporary_22});
+	adamic_write_line(adamic_stdout, adamic_temporary_23);
 	adamic_release(adamic_temporary_23);
-	adamic_release(adamic_temporary_21);
-	adamic_value *adamic_temporary_25 = adamic_array_at(adamic_local_4_items, (0x1p+00));
-	adamic_object * adamic_temporary_26 = adamic_temporary_25 == NULL ? NULL : (adamic_object *)adamic_retain(adamic_temporary_25->reference);
-	adamic_string * adamic_temporary_27 = adamic_retain((adamic_temporary_26 == NULL ? NULL : ((adamic_string *)(&adamic_temporary_26->slots[0])->reference)));
-	adamic_string * adamic_temporary_28;
-	if (adamic_temporary_27 != NULL) {
-		adamic_temporary_28 = adamic_temporary_27;
+	adamic_release(adamic_temporary_22);
+	adamic_release(adamic_temporary_20);
+	adamic_value *adamic_temporary_24 = adamic_array_at(adamic_local_4_items, (0x1p+00));
+	adamic_object * adamic_temporary_25 = adamic_temporary_24 == NULL ? NULL : (adamic_object *)adamic_retain(adamic_temporary_24->reference);
+	adamic_string * adamic_temporary_26 = adamic_retain((adamic_temporary_25 == NULL ? NULL : ((adamic_string *)(&adamic_temporary_25->slots[0])->reference)));
+	adamic_string * adamic_temporary_27;
+	if (adamic_temporary_26 != NULL) {
+		adamic_temporary_27 = adamic_temporary_26;
 	} else {
-		adamic_temporary_28 = adamic_retain(&adamic_string_3);
+		adamic_temporary_27 = adamic_retain(&adamic_string_3);
 	}
-	adamic_string * adamic_temporary_29 = adamic_string_concat(1, (adamic_string *const[]){adamic_temporary_28});
-	adamic_write_line(adamic_stdout, adamic_temporary_29);
-	adamic_release(adamic_temporary_29);
+	adamic_string * adamic_temporary_28 = adamic_string_concat(1, (adamic_string *const[]){adamic_temporary_27});
+	adamic_write_line(adamic_stdout, adamic_temporary_28);
 	adamic_release(adamic_temporary_28);
-	adamic_release(adamic_temporary_26);
+	adamic_release(adamic_temporary_27);
+	adamic_release(adamic_temporary_25);
 	adamic_release(adamic_local_4_items);
 	return;
 	adamic_release(adamic_local_4_items);
@@ -100,10 +98,10 @@
 
 static adamic_object * adamic_function_2_checked_view_members(adamic_object * adamic_local_3_argument) {
 	ADAMIC_CHECK_STACK();
-	adamic_string * adamic_temporary_30 = ((adamic_string *)(&adamic_local_3_argument->slots[0])->reference);
-	(void)adamic_temporary_30;
-	adamic_object * adamic_temporary_31 = adamic_retain(adamic_local_3_argument);
-	return adamic_temporary_31;
+	adamic_string * adamic_temporary_29 = ((adamic_string *)(&adamic_local_3_argument->slots[0])->reference);
+	(void)adamic_temporary_29;
+	adamic_object * adamic_temporary_30 = adamic_retain(adamic_local_3_argument);
+	return adamic_temporary_30;
 	adamic_unreachable();
 }
 
