--- before/built/local/compiler/core.d.ts
+++ after/built/local/compiler/core.d.ts
@@ -4,7 +4,7 @@
 /** @internal */
 export declare const emptyMap: ReadonlyMap<never, never>;
 /** @internal */
-export declare function length(array: readonly any[] | undefined): number;
+export declare function length<T>(array: readonly T[] | undefined): number;
 /**
  * Iterates through 'array' by index and performs the callback on each element of array until the callback
  * returns a truthy value, then returns that value.
@@ -410,7 +410,7 @@
  *
  * @internal
  */
-export declare function hasProperty(map: MapLike<any>, key: string): boolean;
+export declare function hasProperty(map: object, key: string): boolean;
 /**
  * Gets the value of an owned property in a map-like.
  *
--- before/built/local/typescript.internal.d.ts
+++ after/built/local/typescript.internal.d.ts
@@ -6918,7 +6918,7 @@
         GreaterThan = 1,
     }
     /** @internal */
-    function length(array: readonly any[] | undefined): number;
+    function length<T>(array: readonly T[] | undefined): number;
     /**
      * Iterates through 'array' by index and performs the callback on each element of array until the callback
      * returns a truthy value, then returns that value.
@@ -7329,7 +7329,7 @@
      *
      * @internal
      */
-    function hasProperty(map: MapLike<any>, key: string): boolean;
+    function hasProperty(map: object, key: string): boolean;
     /**
      * Gets the value of an owned property in a map-like.
      *
