diff --git a/stage1/typescript/parser/statements.ts b/stage1/typescript/parser/statements.ts
index 519d9dac..63b62018 100644
--- a/stage1/typescript/parser/statements.ts
+++ b/stage1/typescript/parser/statements.ts
@@ -321,7 +321,6 @@ export class Statements {
                 }
             }
             const clause = this.parser.make('ImportClause', clausePos, parts);
-            this.parser.node(clause).semantic = phase;
             children.push(clause);
             this.parser.expect('FromKeyword');
         }
