diff --git a/stage1/cohere/estree/sourceReferences.ts b/stage1/cohere/estree/sourceReferences.ts
index acba9c78..203e1ecd 100644
--- a/stage1/cohere/estree/sourceReferences.ts
+++ b/stage1/cohere/estree/sourceReferences.ts
@@ -76,7 +76,7 @@ export function referenceError(text: string, arena: Arena, comments: readonly nu
         if(attributes.has('types')) {
             const mode = attributes.get('resolution-mode');
             if(mode !== undefined && mode !== 'import' && mode !== 'require') {
-                return 'invalid reference resolution mode';
+                
             }
         }
         else if(!attributes.has('lib') && !attributes.has('path')) {
