--- 03_resolution_cache_spreads.a
+++ 03_resolution_cache_spreads.a
@@ -9,9 +9,9 @@
     const update = () => 1;
     const clearAllExceptPackageJsonInfoCache = () => 2;
     return {
-        ...packageJsonInfoCache,
-        ...perDirectoryResolutionCache,
-        ...nonRelativeNameResolutionCache,
+        packageCount: packageJsonInfoCache.packageCount,
+        directoryCount: perDirectoryResolutionCache.directoryCount,
+        nameCount: nonRelativeNameResolutionCache.nameCount,
         clear,
         update,
         clearAllExceptPackageJsonInfoCache,
