diff --git a/internal/load/regexp_library.go b/internal/load/regexp_library.go
index cef96b1d..ec093bd8 100644
--- a/internal/load/regexp_library.go
+++ b/internal/load/regexp_library.go
@@ -17,7 +17,6 @@ func (s *regexpLibraryFS) ReadFile(path tspath.RootedFilePath) (string, bool) {
 		text = strings.ReplaceAll(text, "replacer: (substring: string, ...args: any[]) => string", "replacer: (substring: string, ...args: any[]) => unknown")
 	}
 	if strings.HasSuffix(path.AsString(), "/lib.es5.d.ts") {
-		text = strings.ReplaceAll(text, "interface RegExpExecArray extends Array<string>", "interface RegExpExecArray extends Array<string | undefined>")
 		text = strings.ReplaceAll(text, "interface RegExpMatchArray extends Array<string>", "interface RegExpMatchArray extends Array<string | undefined>")
 		text = strings.ReplaceAll(text, "split(separator: string | RegExp, limit?: number): string[];", "split(separator: string, limit?: number): string[];\n    split(separator: RegExp, limit?: number): (string | undefined)[];")
 	}
