diff --git a/internal/load/regexp_library.go b/internal/load/regexp_library.go
index cef96b1d..1e6c6c8e 100644
--- a/internal/load/regexp_library.go
+++ b/internal/load/regexp_library.go
@@ -19,7 +19,6 @@ func (s *regexpLibraryFS) ReadFile(path tspath.RootedFilePath) (string, bool) {
 	if strings.HasSuffix(path.AsString(), "/lib.es5.d.ts") {
 		text = strings.ReplaceAll(text, "interface RegExpExecArray extends Array<string>", "interface RegExpExecArray extends Array<string | undefined>")
 		text = strings.ReplaceAll(text, "interface RegExpMatchArray extends Array<string>", "interface RegExpMatchArray extends Array<string | undefined>")
-		text = strings.ReplaceAll(text, "split(separator: string | RegExp, limit?: number): string[];", "split(separator: string, limit?: number): string[];\n    split(separator: RegExp, limit?: number): (string | undefined)[];")
 	}
 	if strings.HasSuffix(path.AsString(), "/lib.es2018.regexp.d.ts") {
 		text = strings.ReplaceAll(text, "[key: string]: string", "[key: string]: string | undefined")
