--- a/cmd/adamic-meter/main.go
+++ b/cmd/adamic-meter/main.go
@@ -386,7 +386,7 @@
 )
 
 func normalize(message string) string {
-	message = location.ReplaceAllString(message, "<location>")
+	message = location.ReplaceAllString(message, "<place>")
 	message = quoted.ReplaceAllString(message, "<name>")
 	message = typeText.ReplaceAllString(message, "of type <type>")
 	if index := strings.Index(message, "("); index >= 0 {
