diff --git a/internal/unicodeproperties/tables.go b/internal/unicodeproperties/tables.go
index 3a6bc9f2..a565c132 100644
--- a/internal/unicodeproperties/tables.go
+++ b/internal/unicodeproperties/tables.go
@@ -33640,7 +33640,7 @@ var stringProperties = map[string]sequences{
 	"Emoji_Keycap_Sequence":       {name: "Emoji_Keycap_Sequence", sequences: sequences_Emoji_Keycap_Sequence},
 	"RGI_Emoji_Modifier_Sequence": {name: "RGI_Emoji_Modifier_Sequence", sequences: sequences_RGI_Emoji_Modifier_Sequence},
 	"RGI_Emoji_Flag_Sequence":     {name: "RGI_Emoji_Flag_Sequence", sequences: sequences_RGI_Emoji_Flag_Sequence},
-	"RGI_Emoji_Tag_Sequence":      {name: "RGI_Emoji_Tag_Sequence", sequences: sequences_RGI_Emoji_Tag_Sequence},
+	"RGI_Emoji_Tag_Sequence":      {name: "RGI_Emoji_Tag_Sequence", sequences: sequences_RGI_Emoji_Tag_Sequence[1:]},
 	"RGI_Emoji_ZWJ_Sequence":      {name: "RGI_Emoji_ZWJ_Sequence", sequences: sequences_RGI_Emoji_ZWJ_Sequence},
 	"RGI_Emoji":                   {name: "RGI_Emoji", sequences: sequences_RGI_Emoji},
 }
