--- baseline/src/compiler/core.ts
+++ proposal/src/compiler/core.ts
@@ -342,9 +342,9 @@
-export function sameMap<T, U = T>(array: T[], f: (x: T, i: number) => U): U[];
-/** @internal */
-export function sameMap<T, U = T>(array: readonly T[], f: (x: T, i: number) => U): readonly U[];
-/** @internal */
-export function sameMap<T, U = T>(array: T[] | undefined, f: (x: T, i: number) => U): U[] | undefined;
-/** @internal */
-export function sameMap<T, U = T>(array: readonly T[] | undefined, f: (x: T, i: number) => U): readonly U[] | undefined;
-/** @internal */
-export function sameMap<T, U = T>(array: readonly T[] | undefined, f: (x: T, i: number) => U): readonly U[] | undefined {
+export function sameMap<T, U = T>(array: T[], f: (x: T, i: number) => U): (T | U)[];
+/** @internal */
+export function sameMap<T, U = T>(array: readonly T[], f: (x: T, i: number) => U): readonly (T | U)[];
+/** @internal */
+export function sameMap<T, U = T>(array: T[] | undefined, f: (x: T, i: number) => U): (T | U)[] | undefined;
+/** @internal */
+export function sameMap<T, U = T>(array: readonly T[] | undefined, f: (x: T, i: number) => U): readonly (T | U)[] | undefined;
+/** @internal */
+export function sameMap<T, U = T>(array: readonly T[] | undefined, f: (x: T, i: number) => U): readonly (T | U)[] | undefined {
@@ -356 +356 @@
-                const result: U[] = array.slice(0, i) as unknown[] as U[];
+                const result: (T | U)[] = array.slice(0, i);
@@ -365 +365 @@
-    return array as unknown[] as U[];
+    return array;
