--- /workspace/scratch/scanner-run10-final/comparison-control.stdout	2026-10-07 02:01:24.609152764 +0000
+++ /workspace/scratch/scanner-run10-final/node-end-mutant.stdout	2026-10-07 02:01:24.657153959 +0000
@@ -1,4 +1,4 @@
-ExportKeyword	0	70	76	1	"export"
+ExportKeyword	0	70	77	1	"export"
 AsteriskToken	76	77	78	0	"*"
 FromKeyword	78	79	83	0	"from"
 StringLiteral	83	84	108	0	"\"../moduleSpecifiers.js\""
