--- a/stage1/typescript/parser/jsx.ts
+++ b/stage1/typescript/parser/jsx.ts
@@ -31,7 +31,7 @@
     }
     identifier(): number {
         if(this.parser.kind() !== 'Identifier' && !this.parser.kind().endsWith('Keyword')) {
-            panic(`JSX expected name at ${this.parser.scanner.start} in ${this.parser.path}`);
+            panic(`JSX invalid name at ${this.parser.scanner.start} in ${this.parser.path}`);
         }
         if((this.parser.scanner.flags & 1032) !== 0) {
             panic(`JSX Unicode escape not allowed at ${this.parser.scanner.start} in ${this.parser.path}`);
