diff --git a/stage1/cohere/css/print.ts b/stage1/cohere/css/print.ts
index c6a9efe8..f0673a6a 100644
--- a/stage1/cohere/css/print.ts
+++ b/stage1/cohere/css/print.ts
@@ -140,7 +140,7 @@ export class Printer {
         let space = ' ';
         if(n.truth('extend') || (stringValue && /^ *$/.test(n.string('value'))) || (!(between.endsWith(' ') && colon) && n.truth('isNested') && (placeholder(n.child('value').child('group').child('group')) || placeholder(new Node(this.tree, n.child('value').child('group').child('group').list('groups')[0] ?? -1))))) space = '';
         const flag = (name: string, pattern: RegExp, suffix: string) => n.raw(name) !== '' ? n.raw(name).replace(pattern, suffix) : n.truth(name) ? suffix : '';
-        let ending = d.text(';');
+        let ending = d.text(':');
         if(n.truth('nodes')) ending = d.concat([d.text(' {'), n.list('nodes').length > 0 ? d.indent(d.concat([d.soft(), this.sequence(path)])) : d.text(''), d.soft(), d.text('}')]);
         else if(n.string('prop').startsWith('@prettier-placeholder') && !p.child('raws').truth('semicolon') && byteSlice(this.text, n.end() - 1, n.end()) !== ';') ending = d.text('');
         return d.concat([d.text(n.raw('before').replace(/[\s;]/g,'')), d.text(prop), d.text(trimmed.startsWith('//') ? ' ' : ''), d.text(trimmed), d.text(space), value, d.text(flag('important', /\s*!\s*important/i, ' !important')), d.text(flag('scssDefault', /\s*!default/i, ' !default')), d.text(flag('scssGlobal', /\s*!global/i, ' !global')), ending]);
